KEP-2839: Add KEP for in-use protection (#2840)
kubernetes/enhancements: keps/sig-api-machinery/2839-in-use-protection/README.md

#### Story 2

A user sets an object X's dependency information for a particular object Y to be protected by this in-user protection controller. For example, when a user knows that `Service` X shouldn't be erroneously deleted while `Deployment` Y exists for his/her particular use case, the user can set such a dependency.

### Notes/Constraints/Caveats (Optional)

- Cross namespace dependency isn't allowed by [garbage collector](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#owners-and-dependents), however, cross namespace dependency will be required for in-use protection controller by the Secret protection use case above,
- Cyclic dependency can be set by user, which potentially makes dead lock,
- In some cases, users really wan't to delete the objects that are considered to be protected in general.
For example, immutable Secrets might need to be deleted while it is used by a Pod to update the value.

> 

Author: From @liggitt  https://github.com/kubernetes/enhancements/pull/2640#issuecomment-887552939

- inverting the relationship direction means that the component responsible for adding/removing the finalizer needs to have perfect visibility to all objects of all types.
- It's also unclear how this avoids races in a quick "create parent, create child, delete parent" scenario

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

- Malicious users can add dependencies to objects that they don't have right to delete and can be prevent such objects from being deleted.

> ## Design Details

Reviewer: This design is very complicated, and I don't think it's necessary. I currently prefer a "lien" system, which is like finalizers except it prevents *beginning* the deletion process rather than just finishing it. This system is easier to implement and more general (works for conditions other than "object exists").

Author: @lavalamp 

Thank you for your comment.

> I currently prefer a "lien" system, which is like finalizers except it prevents beginning the deletion process rather than just finishing it. 

Yes. It is how the current pv-protection and pvc-protection works. So, if the secret-protection is allowed to be implemented in the same way, it will solve my concern.

The background of this KEP is to consider when and to which the finalizer be added (Note that pv-protection and pvc-protection adds their finalizers to all the PVs and PVCs on createion by using admission hook, regardless of whether it is actually used).

As commented [here](https://github.com/kubernetes/enhancements/pull/2640#discussion_r632691232) and [here](https://github.com/kubernetes/enhancements/pull/2640#discussion_r632692560), it is asked to add the finalizer to a resource only when the resource is actually used. I understand that this is for minimizing the impact of breaking the current behavior as commented [here](https://github.com/kubernetes/enhancements/pull/2640#discussion_r632695296), and it makes sense for this purpose.

However, by doing so, there will be a chance that the resource is deleted before the finalizer is added even when it is used.

Garbage collector is taking the approach to add finalizer only when there are owned resources, and it has been working well for a long time. Therefore, I created this KEP in the hope that this approach solves this issue (Also, the codes are rather big for each controller to implement, so I'm proposing this as a generic mechanism). Although, according to [this comment](https://github.com/kubernetes/enhancements/pull/2840#discussion_r677519448), this approach alone won't solve this issue ...

Any ideas or suggestions?

Reviewer: > Yes. It is how the current pv-protection and pvc-protection works.

This is confusing, I'm not sure what you meant; we do not have liens right now. And finalizers don't prevent deletion from beginning.

> add the finalizer to a resource only when the resource is actually used

In a distributed, concurrent database, you have to do a little dance: 1. add the lien; 2. create your item; 3. on failure or on deletion of item, remove the lien.

I maintain that some variation of this dance happens no matter what, and it's best if it's explicit and the client controls it.

(IOW, I consider the proposal recorded here right now to be very racy. GC is also racy in at least one case but there's not simple alternatives.)

Author: > This is confusing, I'm not sure what you meant; we do not have liens right now.

Oh, I now understand what you mean by lien system. Lien is a different concept from finalizer. Finalizer allows beginning the deletion process but prevents finishing it until the finalizer is removed. On the other hand, lien prevents beginning the deletion process itself until the lien is removed. 

I still don't understand what issues are solved by the difference. Could you explain it?
Also, are there plans to implement it?

Author: One of the points seems that the lien is added before the item is created, so there seems no room to erroneously be deleted before the item is found to be created. Then, how we guarantee that the lien is deleted properly and not deleted by others erroneously?

Reviewer: Liens don't exist yet and there aren't currently plans to add them. Given your problem statement, I expected this KEP to be about them :)

Currently deletion is irrevocable, and finalizers just slow the process down. The missing feature is some way to prevent the deletion in the first place. The requirements include that multiple actors must be able to independently preserve an object for different reasons. Liens accomplish this. I don't know if @deads2k has heard of any simpler alternative. (The "lien" name is from @bgrant0607, btw)

> how we guarantee that the lien is deleted properly and not deleted by others erroneously?

Well, I expect a KEP about liens to work that out. :) 

Also, what to do if a namespace is deleted but one of the objects in it has a lien on it?

Author: Thank you very much for your explanation.
It sounds the real solution for this KEP's scope.

Then, I will wait for the lien to be implemented and go back to the original discussion on how to handle the secret specific protection until the lien is implemented.

Reviewer: Just to be clear, no one is working on it, there's not even a KEP.

Author: I mean to work without lien like [pv-protection](https://github.com/kubernetes/enhancements/issues/499) and [pvc-protection](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/postpone-pvc-deletion-if-used-in-a-pod.md), and eventually replace it with using lien once it becomes available.

Users or controllers can add a string to `Liens` to ask to protect a resource.
A newly introduced validation in api-server will reject deletion requests for resources with non-empty `Liens`.

`Liens` is defined as a slice of strings, like `Finalizers`.
The strings need to be namespaced keys.
Multiple users or controllers can add their `Liens` for their own purpose, and they can remove their own `Liens` when it is no longer needed.
`Liens` should be added per controller or per user basis.
Deletion requests of the resource are blocked until the last `Liens` for the resource is removed
(Just to be clear, the difference between `Liens` and `Finalizers` is that `Liens` blocks the deletion request itself, while `Finalizers` blocks the deletion to be completed).

A PoC implementation for in-use protection can be found, [here](https://github.com/mkimuram/kubernetes/commits/lien).
Also, how it can be consumed by Secret protection controller can be found, [here](https://github.com/mkimuram/secret-protection/tree/lien).

### API Changes

`Liens` is added to `ObjectMeta`.

```go
Liens []string
```

> 

Reviewer 2: Can you please add the summary of the comments discussion why we decided with a simple []string instead of more sophistcated struct?

Author: Added.

    - Verify that foreground owner deletion isn't complete while dependent with blockOwnerDeletion=true and lien exists
    - Verify that namespace deletion isn't complete while a resource with lien exists in the namespace
    - Verify that adding liens to non-nil DeletionTimestamp fails
- For Beta, scalability tests are added to exercise this feature.
- For GA, the introduced e2e tests will be promoted to conformance.

### Graduation Criteria

#### Alpha

- Feature implemented behind a feature flag
- Initial e2e tests completed and enabled

#### Beta

- Gather feedback from developers and surveys
- Additional tests are in Testgrid and linked in KEP
- Consider making lien a subresource for giving tighter RBAC

#### GA

- Allowing time for feedback

**Note:** Generally we also wait at least two releases between beta and
GA/stable, because there's no opportunity for user feedback, or even bug reports,
in back-to-back releases.

**For non-optional features moving to GA, the graduation criteria must include
[conformance tests].**

[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md

> #### Deprecation

Reviewer 2: I don't think we're planning any deprecation at this point - let's remove it.

control plane.

For each of these, fill in the following—thinking about running existing user workloads
and creating new ones, as well as about cluster-level services (e.g. DNS):
  - [Dependency name]
    - Usage description:
      - Impact of its outage on the feature:
      - Impact of its degraded performance or high-error rates on the feature:
-->

### Scalability

<!--
For alpha, this section is encouraged: reviewers should consider these questions
and attempt to answer them.

For beta, this section is required: reviewers must answer these questions.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.
-->

###### Will enabling / using this feature result in any new API calls?

Not directly, but users or controllers may call more deletion requests to retry.

###### Will enabling / using this feature result in introducing new API types?

No.

###### Will enabling / using this feature result in any new calls to the cloud provider?

No.

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

Describe them, providing:
  - API type(s): `ObjectMeta`

>   - Estimated increase in size: a slice of `Liens` of size (TBD) B

Reviewer 2: I think we should ensure that the lienID is no longer than N bytes (64? 256?).

The separate question is if we should force upper bound for the size of the slice...

Author: > I think we should ensure that the lienID is no longer than N bytes (64? 256?).

Thank you for suggestion.

Length of each key will be about 32 to 64 bytes.
And the number of keys may be around 4 to 8 at most.

Therefore, 256B will be nice as an estimated size (For most case, it will be 0 to 64B). Updated the KEP with this value.

> The separate question is if we should force upper bound for the size of the slice...

As for current finalizers, the only [validation](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go#L350) seems checking if all the keys are namespaced. So, it seems to have been working well without upper bound until now (I don't mean to say that upper bound isn't necessary just because of it).

If we would like to introduce per key upper bound length, we should add it before beta because it will break compatibility.
As for the upper bound for length of the array or max size of array, I'm not sure that I can set a hard limit in a safe manner (It will introduce inconsistency, because which key can't be added depends on the order of addition).

-->

###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?

<!--
Things to keep in mind include: additional in-memory state, additional
non-trivial computations, excessive access to disks (including increased log
volume), significant amount of data sent and/or received over network, etc.
This through this both in small and large cases, again with respect to the
[supported limits].

[supported limits]: https://git.k8s.io/community//sig-scalability/configs-and-limits/thresholds.md
-->

### Troubleshooting

<!--
This section must be completed when targeting beta to a release.

The Troubleshooting section currently serves the `Playbook` role. We may consider
splitting it into a dedicated `Playbook` document (potentially with some monitoring
details). For now, we leave it here.
-->

###### How does this feature react if the API server and/or etcd is unavailable?

Deletion won't also happens if the API server and/or etcd are unavailable for all controllers.
Therefore, it won't affect the protection.

###### What are other known failure modes?

> <!--

Reviewer 2: The one failure mode that comes to my mind is that users will be leaving the objects iwth `liens` and as a result all our GC machinery (namespace deletion, TTL-after-finish, pod-gc, and so on) will not be able to delete those object, eventually leading to filling up etcd...

Author: It could also happen in current GC, if the other finalizers are not removed for a long time. I will add this as a failure mode, if it is still worth describing here.

- Protect `PersistentVolumeClaim` while being used by `Pod` (Replace the background logic for pvc-protection)
- Protect `Secret` while being used by `Pod`, `PersistentVolume`, `VolumeSnapshotContent` (Secret protection above)
- Protect `ConfigMap` while being used by `Pod`
- Protect `StorageClass` while being used by `PersistentVolume`
- Protect `VolumeSnapshot` while being used by `VolumeSnapshotContent`
- Protect `VolumeSnapshotClass` while being used by `VolumeSnapshotContent`
- Protect `User`, `Group`, `ServiceAccount` and `Role` from `RoleBinding`
- Protect `User`, `Group`, `ServiceAccount` and `ClusterRole` from `ClusterRoleBinding`
- Protect resources that users marked as in-use

### Goals

- Provide a generic mechanism that prevents a resource from being deleted when it shouldn't be deleted

### Non-Goals

- Provide a specific mechanism to decide which particular objects should be prevented from being deleted when other particular objects exist
- Provide a generic mechanism that prevents a resource from being updated when it shouldn't be updated

## Proposal

A new field `Liens` to mark the object not to be deleted is introduced in object Metadata.

> Deletion requests for resources with non-empty `Liens` will be blocked by a newly introduced validating admission webhook.

Reviewer: webhook is good for prototyping, for real use we would build this directly into the deletion strategy, probably.

### User Stories (Optional)

#### Story 1

Another controller to protect Secret from deletion watches all Secrets and their potential user objects, like `Pod`, `PersistentVolumes`, and `VolumeSnapshotContent`. Once it find the Secret is used by one of the objects, it updates the Secret's `Liens` and ask this mechanism to block deletion request for the secret.

#### Story 2

A user sets `Liens` to an object to be protected by this mechanism. For example, when a user knows that `Namespace` X shouldn't be erroneously deleted while `Deployment` Y exists for his/her particular use case, the user can set such a protection.

### Notes/Constraints/Caveats (Optional)

- In some cases, users really wan't to delete the objects that are considered to be protected in general.

> For example, immutable Secrets might need to be deleted while it is used by a Pod to update the value.

Reviewer: Since liens are not ACL'd, they just need to be manually removed to delete the object.

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

Most of the design ideas come from [here](https://github.com/kubernetes/kubernetes/issues/10179).
Users or controllers can add a string to `Liens` to ask to protect a resource.

> A newly introduced validating admission webhook will reject deletion requests for resources with non-empty `Liens`.

Reviewer: A prior comment applies to this line too.

`Liens` is defined as a slice of strings, like `Finalizers`.
Multiple users or controllers can add their `Liens` for their own purpose, and they can remove their own `Liens` when it is no longer needed.
Deletion requests of the resource are blocked until the last `Liens` for the resource is removed
(Just to be clear, the difference between `Liens` and `Finalizers` is that `Liens` blocks the deletion request itself, while `Finalizers` blocks the deletion to be completed).

> To make multiple users and controllers share the `Liens` field properly, they should include their identity to the `Liens` string and they should only delete their own `Liens` strings.

Reviewer: I think there needs to be a debate about the structure. String list is ok, but maybe we can do better. A key-value pair or struct of some sort might be useful for making the owner more obvious.

Reviewer 3: I'm not so convinced here - consistency with finalizers is nice, and is it really confusing?

Reviewer 2: Thinking a bit more about that - clearly a struct would be more extensible.
But in this particular case, I'm not convinced it's actually an advantage.

I already commented below on scalability implications if we decide to store Lien per object - I don't think it's a good solution. And making this extensible would be definition be prone to those kind of regressions.

So my personal preference is actually towards a simple string (which BTW as Tim pointed out again would be consistent with Finalizers which is pretty nice).

Author: Change back to slice-of-strings Liens and deleted this line.

A PoC implementation for in-use protection can be found, [here](https://github.com/mkimuram/kubernetes/commits/lien).
Also, how it can be consumed by Secret protection controller can be found, [here](TBD).

### API Changes

`Liens` is added to `ObjectMeta`.

```
Liens []string
```

> ### Other Design Considerations

Reviewer: Missing: what happens if you delete a namespace and some objects within it have liens on them?

Author: Maybe it is also good to describe how lien works with `ownerReference`.

Differences of behaviors between "Foreground cascading deletion" and 
"Background cascading deletion" is expected to be a bit tricky. And behaviors of Namespace deletion will be like "Foreground cascading deletion" with `ownerReference.blockOwnerDeletion=true`.

#### Opt-out Protection

In some cases, users may need to opt-out protection to force delete the resrouce.
However, controllers may regularly add `Liens` to a resource, therefore, just removing `Liens` before force delete won't be enough.
To avoid such a situation, below opt-out mechanism is provided.

> - If `k8s.io/skip-lien` annotation is added to the resource, the resource is regarded as it has empty `Liens`. Therefore, the resource is deleted regardless of its `Liens` field.

Reviewer: I don't think we can do it this way. I'd rather have a query parameter on delete permitting you to ignore them. But I'm not sure if this problem needs to be solved at all, since if you delete the thing anyway you're going to make whatever controller it is very angry and that's probably not a good idea.

#### Block Adding Additional `Liens` while Deleting

`Liens` shouldn't be added after `DeletionTimestamp` is non-nil, which means pre-deletion processes for finalizers are being handled.
Adding `Liens` while deleting itself won't do any harm for the cluster, because the resource will be deleted without any additional deletion requests that will be blocked by `Liens`.
However, users may think that the successful addition of the `Liens` means that the resource won't be deleted until the `Liens` are deleted, which isn't true.
To avoid such a misunderstanding from users, a request for adding additional `Liens` to a resource with non-nil `DeletionTimestamp` should be blocked by validating admission webhook for update.

#### Cross Namespace Dependency Consideration

`Liens` can be added only by users or controllers that have create/update rights to resources.
Therefore, this mechanism itself doesn't have issues for cross namespace dependency.
However, when it is used with controllers that have access across namespaces,
they can easily introduce a cross namespace dependency.
For example, adding `Liens` to resource X in namespace x, because resource Y exists in namespace y.

> This may allow malicious users to block the resources in other namespace from being deleted.

Reviewer: I don't buy this, if you have update permission in your namespace you can delete the thing anyway.

To minimize issues caused by a cross namespace dependency, it should be considered to implement
a mechanism to opt-in/opt-out adding liens across namespace in such controllers.

> For example, a controller may choose to add `Liens` only to resources that has `example.com/sample-controller/allow-cross-namespace-liens: true` annotation, if the dependency resource isn't in the same namespace.

Reviewer: I would recommend removing this whole section, honestly. I think it's a separate discussion for each controller and I don't think this KEP should recommend a general solution; I'm not convinced there's a problem at all.

On the other hand, if `Liens` are added per reason basis, such a controller can remove the `Lien` for the resource, without considering other resources.
Also, the latter way helps users to know which particular resources are using the resource.

`Liens` allows both per controller `Liens` and per reason `Liens`.
Only requirements are to include identity that who add it and to only remove its own's.

Note that examples of per controller `Liens` and per reason `Liens` are as follows:

Let's assume that a secret protection controller would like to protect Secret `my-secret` in `ns1` when
both PersistentVolume `my-pv1` and Pod `my-pod1` in namespace `ns2` are using it.

- Per Controller `Liens`:
```
apiVersion: v1
kind: Secret
metadata:
  namespace: ns1
  name: my-secret
  liens:
    - "k8s.io/secret-protection-controller"
```

- Per Reason `Liens`:
```
apiVersion: v1
kind: Secret
metadata:
  namespace: ns1
  name: my-secret
  liens:
    - "k8s.io/secret-protection-controller v1/PersistentVolume//my-pv1"

>     - "k8s.io/secret-protection-controller v1/Pod/ns2/my-pod1"

Reviewer: k/v pairs would make this much prettier. Or a list of structs with {owner, reason} fields. etc.

GA/stable, because there's no opportunity for user feedback, or even bug reports,
in back-to-back releases.

**For non-optional features moving to GA, the graduation criteria must include
[conformance tests].**

[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md

#### Deprecation

- Announce deprecation and support policy of the existing flag
- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
- Address feedback on usage/changed behavior, provided on GitHub issues
- Deprecate the flag

### Upgrade / Downgrade Strategy

<!--
If applicable, how will the component be upgraded and downgraded? Make sure
this is in the test plan.

Consider the following in developing an upgrade/downgrade strategy for this
enhancement:
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to maintain previous behavior?
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to make use of the enhancement?
-->
- Upgrade:

>   - When lien admission plugin is not enabled in the cluster, object deletion won't be prevented by this feature.

Reviewer: If we do this in core, we can not make this optional IMO

- Provide a generic mechanism that prevents a resource from being updated when it shouldn't be updated
- Implement any of the above potential use cases

## Proposal

A new field `Liens` to mark the object not to be deleted is introduced in object Metadata.
Deletion requests for resources with non-empty `Liens` will be blocked by a newly introduced validation in api-server.

### User Stories (Optional)

#### Story 1

Another controller to protect Secret from deletion watches all Secrets and their potential user objects, like `Pod`, `PersistentVolumes`, and `VolumeSnapshotContent`. Once it find the Secret is used by one of the objects, it updates the Secret's `Liens` and ask this mechanism to block deletion request for the secret.

#### Story 2

A user sets `Liens` to an object to be protected by this mechanism. For example, when a user knows that `Namespace` X shouldn't be erroneously deleted while `Deployment` Y exists for his/her particular use case, the user can set such a protection.

<!--
### Notes/Constraints/Caveats (Optional)
-->

### Risks and Mitigations

> <!--

Reviewer 3: I think @deads2k comment regarding races is a good point for "risks"

Author: Updated.

In other words, users can delete its namespace or other objects in the namespace.
The point that needs to be considered is whether the namespace is actually deleted, or whether its deletion completes immediately.
On namespace deletion, all the objects in the namespace are deleted, and the deletion of the namespace isn't completed until all the objects are deleted.
As a result, deletion of the namespace isn't completed until the object is deleted.

#### Block Adding Additional `Liens` while Deleting

`Liens` shouldn't be added after `DeletionTimestamp` is non-nil, which means pre-deletion processes for finalizers are being handled.
Adding `Liens` while deleting itself won't do any harm for the cluster, because the resource will be deleted without any additional deletion requests that will be blocked by `Liens`.
However, users may think that the successful addition of the `Liens` means that the resource won't be deleted until the `Liens` are deleted, which isn't true.
To avoid such a misunderstanding from users, a request for adding additional `Liens` to a resource with non-nil `DeletionTimestamp` should be blocked by api-server.

> #### Per Controller `Liens` vs Per Reason `Liens`

Reviewer 3: I'm unclear - if I interpret @detiber correctly (and if not, sorry) he's describing a dep-graph, where deletion of a lien-holder is sufficient to clear a lien.

That's SIGNIFICANTLY more involved (I think) than a simple lien.

Am I wrong?

> #### Per Controller `Liens` vs Per Reason `Liens`

Reviewer 4: The ideal scenario I had in mind would indeed require a dependency graph, but I still think it could be approximated using a common controller or shared library as @mkimuram describes.

We've seen various iterations of this within the Cluster API project for a few different types of scenarios.

The main scenario we've hit is ensuring proper cleanup of resources and centralized reporting of status.  We don't want folks to have to know the proper order of deletion of specific resources to ensure a cluster is cleanly deleted, so we encourage folks to delete the Cluster resource rather than the other related resources, but it would be nice to have a way to avoid users potentially inadvertently getting things wedged by deleting things in the wrong order or accidentally deleting something they didn't mean to. Today we accomplish this mostly through documentation, and implementing deletion of "child" resources through the higher level controllers and try to ensure that we don't introduce false assumptions and bugs if deletion order is done "improperly" or if foreground deletion is used

The other use case we've hit is around introducing safety around managed child resources. Take the example of a KubeadmControlPlane, this resource manages multiple Machine and related resources that make up the control plane of the kubernetes clusters being managed. If a user unintentionally deletes managed "child" resources that "belong" to a KubeadmControlPlane, then they could potentially not only destroy their cluster, but destroy the underlying data it contained. However, if the user is already deleting the KubeadmControlPlane resource, then there is no need to provide that level of protection. KubeadmControlPlane is the obvious case we have here for Cluster API, but similar behavior could be beneficial to other "parent"/"child" CRD/controller relationships as well.

To avoid such a misunderstanding from users, a request for adding additional `Liens` to a resource with non-nil `DeletionTimestamp` should be blocked by api-server.

#### Format of `Liens`

As a result of discussions, format of `Liens` is decided as a slice of strings and it needs to be used per controller(per user) basis.
The main reasons are scalability and better aligned with `Finalizers`.

- Per controller(Per user) `Liens`
  - Pros:
    - Aligned with `Finalizers`
    - Good scalability (Size of the field and number of API calls required)
  - Cons:
    - Bad UX (Users won't know which other objects are actually using the object)
    - Complex consuming controllers' implementation (Each controller needs to track the using objects)
- Per reason(Per object) `Liens`
  - Pros:
    - Good UX (Users can know which other objects are actually using the object)
    - Simple consuming controllers' implementation (Lien tracks the using objects)
  - Cons:
    - Not aligned with `Finalizers`
    - Bad scalability (Size of the field and number of API calls required)

> #### Race of removing/adding liens

Reviewer 2: > OTOH if you want a runtime referential integrity feature, then we have to permission the addition and removal of liens to make sure the wrong parties can't break the referential integrity. This sounds very complicated to me and I don't really want to do it if we don't have to.

@lavalamp - I'm somewhat on the fence here. I agree that "ability to ignore existing liens" is a very simple mitigation to the problem, and that we need to allow certain users to ignore liens. However, ideally it would be great if we could limit the exact set of users that can do that.

That said - thinking more about it, I guess it's solvable. I can create a ValidatingWebhook that via admission request gets the DeleteOptions (I'm assuming that the "NotBlockOnLien" (or however we call it) option will be added to DeleteOptions) and it can allow/reject the request based on UserInfo.
So it seems that technically it is possible to achieve what I need.
Or am I missing something above? [If not - I'm onboard with the currently proposed mitigation.]

@deads2k

Author: Applied @wojtek-t 's idea to the KEP.

Reviewer: I hadn't thought about doing it in a webhook, that would make it optional functionality. We could get very similar functionality by e.g. adding an RBAC verb of "ignoreliens", and checking that verb when someone attempts to use the DeleteOption. I wonder if this would be acceptable to @deads2k? I think it's much simpler than controlling who is allowed to add the liens.

Reviewer 2: Modifying RBAC would of course work too, but I had an impression that you're both against that.
If that's not the case - I think that would be a preferred option...

Author: Just for your information, I've implemented a prototype of force delete [here](https://github.com/mkimuram/kubernetes/commits/lienv2-forcedel) (Only the last 5 commits are for force delete). It only allows users in "system:masters" group to specify `NotBlockOnLien` delete option, which can be specified via `--not-block-on-lien=true` flag in kubectl delete command. I hope that this prototype helps further discussion.

Please also see the console log below:

<details>

- With a user in "system:masters" group
```
kubectl create secret generic test-secret --from-literal='username=my-app' --from-literal='password=39528$vdg7Jb'
kubectl patch secret test-secret -p '{"metadata":{"liens":["never.delete/me"]}}' --type=merge

cluster/kubectl.sh delete secret test-secret
Error from server (Forbidden): secrets "test-secret" is forbidden: deletion not allowed by liens

cluster/kubectl.sh delete --not-block-on-lien=true secret test-secret
secret "test-secret" deleted
```

- With a user __not__ in "system:masters" group (the test user, or mynamespace-user, is created by [this way](https://jeremievallee.com/2018/05/28/kubernetes-rbac-namespace-user.html))
```
kubectl create secret generic test-secret --from-literal='username=my-app' --from-literal='password=39528$vdg7Jb'
kubectl patch secret test-secret -p '{"metadata":{"liens":["never.delete/me"]}}' --type=merge

cluster/kubectl.sh delete secret test-secret
Error from server (Forbidden): secrets "test-secret" is forbidden: deletion not allowed by liens

cluster/kubectl.sh delete --not-block-on-lien=true secret test-secret
Error from server (Forbidden): secrets "test-secret" is forbidden: specifying NotBlockOnLien=true delete option is not allowd to user system:serviceaccount:mynamespace:mynamespace-user
```

</details>

Author: Added this to the KEP as alpha's scope. Note that kube-apiserver's option part of the current description isn't implemented in the prototype, so I'm not sure that it really works (Also, please check if setting `system:controller:namespace-controller` user to default will allow namespace lifecycle controller to force delete).

A cluster-admin would be racing to update the object to remove the lien and delete the object before a namespace editor is able to place the lien back. As a result, a namespace admin or editor can create an object that a cluster-admin or namespace lifecycle controller cannot delete, which shouldn't be allowed.

> To mitigate the risk, a command line option to force delete a resource with liens will be added.

Reviewer 2: Technically - it's not a command-line option, but rather API option.
It should probably land in DeleteOptions:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L482

As a result, other controllers will execute their pre-deletion actions while a finalizer for protection still exists.

This KEP aims to provide a generic mechanism for protection that solves the above issues with Finalizers, because such use cases are also found for protecting secrets and configmaps in [KEP-2639](https://github.com/kubernetes/enhancements/pull/2640).

Potential use cases are:
- Protect `PersistentVolume` while being used by `PersistentVolumeClaim` (Replace the background logic for pv-protection)
- Protect `PersistentVolumeClaim` while being used by `Pod` (Replace the background logic for pvc-protection)
- Protect `Secret` while being used by `Pod`, `PersistentVolume`, `VolumeSnapshotContent` (Secret protection above)
- Protect `ConfigMap` while being used by `Pod`
- Protect `StorageClass` while being used by `PersistentVolume`
- Protect `VolumeSnapshot` while being used by `VolumeSnapshotContent`
- Protect `VolumeSnapshotClass` while being used by `VolumeSnapshotContent`
- Protect `User`, `Group`, `ServiceAccount` and `Role` from `RoleBinding`
- Protect `User`, `Group`, `ServiceAccount` and `ClusterRole` from `ClusterRoleBinding`

> - Protect resources that users marked as in-use

Reviewer: Mention: for accidental deletion protection

Author: Added.

### Goals

> - Provide a generic mechanism that prevents a resource from being deleted when it shouldn't be deleted

Reviewer: Decide: is this referential integrity, or just advisory. (I vote the latter)

Author: Added as advisory.

### Non-Goals

- Provide a specific mechanism to decide which particular objects should be prevented from being deleted when other particular objects exist
- Provide a generic mechanism that prevents a resource from being updated when it shouldn't be updated
- Implement any of the above potential use cases

## Proposal

> A new field `Liens` to mark the object not to be deleted is introduced in object Metadata.

Reviewer 5: I hope we can find a more human friendly name :)

UserFinalizer or ActiveUsage or or Utilizers

Author: Actually the name came from [here](https://github.com/kubernetes/kubernetes/issues/10179#issuecomment-580932333), but `Utilizers` looks nice to me, as a non-native speaker of English. I will wait for feedback from others (I think that sig-api-machinery has been called this concept as Liens for a long time, so they may have their reasons for the name).

Reviewer: "Lien" is a really tight analogy to what this is doing, which is probably why @bgrant0607 chose the word initially. Everyone with a mortgage should understand what a lien is; that's a lot of people, though certainly not everyone.

I personally hate the word "utilize" because we already have the perfectly good "use", but unfortunately "users" would not be very clear.

If we can't use "lien" the next word on my list would probably be "Reservations" or "Reservers".

Author: > Everyone with a mortgage should understand what a lien is; that's a lot of people, though certainly not everyone.

It is exactly what "as a non-native speaker of English" meant. Actually, I used dictionary when I first heard the name of the concept. Although, it might be just due to the lack of my vocabulary.

Reviewer 6: I didn't have any trouble understanding lien in this context.  Users is likely to be confused with "people using kube".  Utilizers seems inaccurate because this is not necessarily tracking the actual resource which has declared, "don't delete this"

Reviewer 7: How about "InUse"?

Reviewer 3: I like the word "liens" for this - it's very specific and maps very closely to the meaning.

if we NEED to go chattier, the next most common phrase I hear is "deleteInhibitors".  That's a lot more letters for no additional connotation.

It's rare that we find a "just right" word.

Reviewer 5: The potential for confusion around `lein` in casual usage (homonyms), it's relatively niche usage in a business context (I find the mortgage argument unconvincing simply because I bet most homeowners *don't* think of the word "liens"), and potentially negative connotations around the word (although I rate this the least significant).

I was hoping that some of the justification for the name would come along with the KEP, vs being left as a blank space for me to guess at what else has been tried :)

The most succinct human definition I've seen so far is:

> A consumer of this resource considers its removal to have a significant negative impact on correct operation, and furthermore desires that any other human or automation interacting with the resource NOT delete it unless said client is fully aware of the implications (which roughly revolve around either major disruption or an unrecoverable state)

Is that a fair summation?

It doesn't necessarily imply the resource is in use, and I agree that "users" is also potentially confusing from the perspective of humans vs machines.  It does argue that a particular client wants to add an additional level of importance to a resource to distinguish between "replaceable automatically without major impact" (typically implied by the controller field) vs "not replaceable automatically or without major impact".

Some other terms that potentially might make sense (would like to see arguments for/against for each):

* `hold`
* `lease`
* `claim`

Lein is typically defined as "hold on property", and I agree that this is very much a hold on something that is the property (owned by) a human or machine.  It is not "controlled", and I don't think the word specifically here is "ownership".

Reviewer 8: `deletionBlockers`. Everybody will understand that.

As a non-native speaker I have never heard "liens".

Reviewer 2: As non-native speaker, I've never heard "liens" either.
But I must admit that after translating it back then when I first seen that, I actually like this

Reviewer 9: As another non-native I'm in favor of `liens` as well, `deletionBlockers` is pretty mouthful and the former translates pretty well.

Reviewer 3: I'm thinking of a number between 0 and 3^32.  Whoever guesses closest gets to choose the name from the following list:

1) liens
2) deletionBlockers
3) holds
4) deleteInhibitors

Go.

Reviewer 3: I'm kidding of course.  I still like "liens" but I can live with "holds".

Reviewer 10: Chipping in from SIG Docs. To help us decide, would someone like to draft an entry in the Kubernetes glossary that includes an initial definition of “lien”?

Reviewer 3: Lien: An indication that some entity is relying on an API object.  Deletion of objects with liens will be prevented until the interested party releases the lien.

Deletion requests for resources with non-empty `Liens` will be blocked by a newly introduced validation in api-server.

### User Stories (Optional)

#### Story 1

Another controller to protect Secret from deletion watches all Secrets and their potential user objects, like `Pod`, `PersistentVolumes`, and `VolumeSnapshotContent`. Once it find the Secret is used by one of the objects, it updates the Secret's `Liens` and ask this mechanism to block deletion request for the secret.

#### Story 2

A user sets `Liens` to an object to be protected by this mechanism. For example, when a user knows that `Namespace` X shouldn't be erroneously deleted while `Deployment` Y exists for his/her particular use case, the user can set such a protection.

> 

Reviewer: add an accidental deletion protection user story? E.g. User knows this is an important object, and that e.g. terraform likes to delete and recreate objects, so the user sets a lien so that an accidental deletion will fail.

Author: Replaced the story 2 with it (because references to the using objects are no longer managed by lien, so the old one is covered by the new one).

<!--
### Notes/Constraints/Caveats (Optional)
-->

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

Most of the design ideas come from [here](https://github.com/kubernetes/kubernetes/issues/10179).
Users or controllers can add a string to `Liens` to ask to protect a resource.
A newly introduced validation in api-server will reject deletion requests for resources with non-empty `Liens`.

`Liens` is defined as a slice of strings, like `Finalizers`.
The strings need to be namespaced keys.

> Multiple users or controllers can add their `Liens` for their own purpose, and they can remove their own `Liens` when it is no longer needed.

Reviewer 6: Are liens modified using spec or status or a different subresource?  When we created finalizers, we realized that the power to change a finalizer did not necessarily imply the power to change spec or status, but we never formalized that. 

I'd like to avoid repeating the same mistake with another global metadata field.  The power to create and modify liens seems independent of the ability to modify spec or status.

Author: In the current KEP, advisory feature model is chosen, so Liens field remains in metadata and is not defined as a subresource. If we find use cases that require hard locking feature and choose the model, making it a subresource and moving it to outside the metadata would be needed.

Reviewer 6: > In the current KEP, advisory feature model is chosen, so Liens field remains in metadata and is not defined as a subresource.

That may not be sufficient.  Someone may want to grant a controller sufficient power to set or clear a lien, but not the power to modify the intent of the resource or the status.

Reviewer: I believe creation of a subresource to allow an actor to change liens but not other parts of the object is possible as a followup. I am not sure it's mandatory for the first version.

I believe the right to edit spec should automatically include the right to edit liens, to support creating objects with liens already in place.

Reviewer 6: > I am not sure it's mandatory for the first version.

Having been through finalizers, if it's not in the first version, I lack confidence that it will appear later.

Author: xref: https://github.com/kubernetes/kubernetes/issues/72637#issuecomment-630586536

Reviewer 3: Can we pleaaaaaaaaase add a `finalize` subresource and a protocol to add/clear those?

Reviewer 2: > Can we pleaaaaaaaaase add a finalize subresource and a protocol to add/clear those?

+1

Author: @deads2k @lavalamp @thockin @wojtek-t @alvaroaleman Created a separate [KEP](https://github.com/kubernetes/enhancements/pull/3286) for subresourcing finalizers. Please review it.

(Also, I'm updating this KEP).

Reviewer 11: > Would we be able to add the feature today to finalizers or is there a design reason that makes it impossible (backward incompatibility)?

https://github.com/kubernetes/kubernetes/issues/59591 suggests we wouldn't be able to, at least not by default (which is what matters).

`Liens` should be added per controller or per user basis.
Deletion requests of the resource are blocked until the last `Liens` for the resource is removed
(Just to be clear, the difference between `Liens` and `Finalizers` is that `Liens` blocks the deletion request itself, while `Finalizers` blocks the deletion to be completed).

A PoC implementation for in-use protection can be found, [here](https://github.com/mkimuram/kubernetes/commits/lien).
Also, how it can be consumed by Secret protection controller can be found, [here](https://github.com/mkimuram/secret-protection/tree/lien).

### API Changes

`Liens` is added to `ObjectMeta`.

```go
Liens []string

> ```

Reviewer: State the validation criteria, e.g. permitted format of the strings, and the maximum number of liens.

Author: Added format of the lien and maximum number of each lien, but maximum number of liens isn't added to validation as explained.

If a lien is set to Replicaset B, only a deletion request to Replicaset B is blocked.
Users can still request to delete Deployment A and Pod C directly.
When deletion of the Deployment A is requested, whether it completes immediately or not depends on the cascading policy and Replicaset B's `blockOwnerDeletion`.
The behaviors are as follows:

- Foreground cascading deletion:
  - `blockOwnerDeletion=true`: Deletion of Deployment A isn't completed until Replicaset B is deleted
  - otherwise: Deployment A is deleted immediately
- Background cascading deletion: Deployment A is deleted immediately

#### Namespace Deletion

It may also not be clear how [namespace deletion](https://kubernetes.io/docs/tasks/administer-cluster/namespaces/#deleting-a-namespace) behaves with lien.
Therefore, this section describes the behavior.

In summary, lien will block the namespace deletion, but would not block deletion of any resources in the namespace individually.

If a lien is set to a namespace, only a deletion request to the namespace is blocked.
Users can still request to delete each object in the namespace directly, but they can't request to delete it through the namespace deletion.

> If a lien is set to one of objects in a namespace, only a deletion request to the object is blocked.

Reviewer 6: If liens are only advisory and can be skipped, the deletion of a namespace (which requires higher privileges) seems like a valid use of "force delete this item".

Author: I notice that making namespace deletion automatically force delete all resources inside namespace may conflict with my use case described [as an example of the summary](https://github.com/kubernetes/enhancements/blob/7a3d79cb9a3972109fa3de3434b6c4e8c9004870/keps/sig-storage/2639-secret-configmap-protection/README.md#summary) of secret protection KEP. In the use case, lien is used to force right deletion order (Keeping liens while a resource is used by other resources and removing liens when it is no longer used). So, just making force delete will make deletion order inconsistent.  

Making deletion 2 steps, (1) normal delete and (2) force delete, may resolve this concern. Does it make sense?

Author: As for "proper order of deletion of specific resources", @detiber 's [comment](https://github.com/kubernetes/enhancements/pull/2840#discussion_r704695168) seems a good description (I'm referencing it here to show deletion order use case is a common use case). It is talking about how lien will be consumed, but it may be good to describe it in this KEP.

Reviewer 6: > Making deletion 2 steps, (1) normal delete and (2) force delete, may resolve this concern. Does it make sense?

I'm open to having a multi-pass namespace cleanup controller.  Under what conditions would the namespace cleanup controller force delete?

Author: I'm thinking about force deleting if all the errors are due to lien [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/namespace/deletion/namespaced_resources_deleter.go#L133).

We need to consider a bit more like:
- Wait to see if lien on the resources may be eventually removed due to the using resources are deleted, before actually start force deleting (Just sleep?)
- Improve normal deletion logic to avoid deleting resource with lien (and start force deleting when no error there)

But the basic idea will be something like that.

Author: Also, this behavior is added to the KEP.

The main reasons are scalability and better aligned with `Finalizers`.

- Per controller(Per user) `Liens`
  - Pros:
    - Aligned with `Finalizers`
    - Good scalability (Size of the field and number of API calls required)
  - Cons:
    - Bad UX (Users won't know which other objects are actually using the object)
    - Complex consuming controllers' implementation (Each controller needs to track the using objects)
- Per reason(Per object) `Liens`
  - Pros:
    - Good UX (Users can know which other objects are actually using the object)
    - Simple consuming controllers' implementation (Lien tracks the using objects)
  - Cons:
    - Not aligned with `Finalizers`
    - Bad scalability (Size of the field and number of API calls required)

#### Race of removing/adding liens

A cluster-admin would be racing to update the object to remove the lien and delete the object before a namespace editor is able to place the lien back. As a result, a namespace admin or editor can create an object that a cluster-admin or namespace lifecycle controller cannot delete, which shouldn't be allowed.

> To mitigate the risk, a new API option in `DeleteOptions` to force delete a resource with liens will be added.

Reviewer 5: At the heart of this feature is the fact that it breaks all existing deletion flows the moment a user adds a lien and the client tool is not aware of it.

That’s a very big change - tools will not work, libraries will not work, etc.

Given our commitment to backward compatibility, I’m really not comfortable with breaking all client deletions.  Instead, I believe liens must be precondition for delete, not a “on by default behavior” (ie Precondition - no leins from specfic tools).

If there is more justification for “break api by default” it needs to be more prominent here (or i’m blind) but I don’t think what is here meets the bar for such a massive disruption.

Author: > Given our commitment to backward compatibility, I’m really not comfortable with breaking all client deletions. Instead, I believe liens must be precondition for delete, not a “on by default behavior” (ie Precondition - no leins from specfic tools).

Could you explain a bit more about your concern and idea to solve it?

Are you worrying about adding Lien itself because it breaks the existing behavior or worrying about adding an option to ignore Lien because it breaks the behavior of Lien?

Reviewer: @smarterclayton 

I think at least for the accidental deletion use case, any thorough mechanism has the property that it breaks workflows deleting the protected object. So you are effectively arguing against thorough mechanisms, e.g. that deletion protection should only affect kubectl or opt-in clients. I think this is not right (e.g. a motivator is saving an object when you forget to turn on Terraform's deletion protection). Since your argument is not against this particular mechanism, I think you have to argue against the use case.

So: for the accidental deletion protection use case, preventing deletion flows is the point and I think that's good, I expect it to be somewhat rare.

Elsewhere in the KEP it's decreed that namespace deletion will always force, but garbage collection should not force. I do not think there's a reasonable design that lacks the "ignore liens" permission. We could also have an "add/update liens" permission, but since at least some users would have that permission, at best that just reduces the number of protected objects, and therefore I don't think it's necessary if important actors can force.

I agree that if you do something stupid like protect all of your pods, the pod-gc controller won't clean them up and your cluster will fill up and stop working. IMO misuses are possible all over the system and while it's arguably our responsibility to make misuse inconvenient or at least not easy, it's not our responsibility to make it impossible; that standard rules out too many useful mechanisms.

Now, in terms of automated application of liens, I think you have a great point, and I think we should debate that separately at each proposed use. This KEP just makes the mechanism possible; we definitely would have to be responsible when using it in an automated fashion. E.g. I am actually not sure that the proposed application for keeping secrets around when a pod uses them is a good one, it sounds risky. But I'm fairly sure that the PVC/PV relationship could benefit from liens.

Reviewer 6: > breaks all existing deletion flows the moment a user adds a lien and the client tool is not aware of it.

Which deletion flow breaks in particular?  Delete requests can fail for any number of reasons.  Since clients already have to handle deletion failures and since the semantics of what happens on a successful delete call does not change, the change will be limited to non-humans (humans could read the message and add --no-really--force-this-time).

The intended application of this approach is to prevent that non-human from taking an action that shouldn't happen (otherwise there wouldn't be a lien).

We're now constrained down to non-human clients that, when operating correctly, should override the expressed desire to avoid deleting a resource.  That is a fairly constrained set.

Reviewer 5: @lavalamp i can buy the argument that that “use of leins” is what is breaking compatibility, in which case i’d be fine if this KEP articulated that and specified that automated leins are what take the burden of proving api compatibility, and thus a kep introducing them has to pass that bar.

Reviewer 12: Question for breakage: is there some way that leins can be overridden for automated tools *without* needing to change the code of the tools?  I'm thinking administrator action, but done outside what may be a third-party tool. If the answer is yes, then there's no compatibility problem.

If the answer is no, it becomes like deleting a table in Postgres 9, where you literally had to kick all users out of the database before you could drop a popular table (or its primary key index).  That's not a good workflow; sometimes you do need to delete things even though they are in use.

Reviewer: The first thing that comes to mind is to add an "auto-override-liens" permission which could be given to particular automated tools.

Before adding that or something similar to the design, I personally would like to hear from more users first that yes, they want some automated tooling to ignore deletion protection. Because I expected people to mostly want protection *from* the automated tooling.

The next step on this is "I want object A to be deletable by controller X even if A is protected. However I want object B to be super-protected so that not even controller X can delete it. However controller Y should still be able to delete object B in that case. I also have object C which should not be deletable by X or Y, but controller Z may delete it. ..." It is easy to model this as an integer instead of as a boolean (e.g. put a number on both the lien and the permission) but I really hope it's not necessary.

Reviewer 12: @lavalamp I'm thinking of cases that go like this:

1. Deploy #3721 failed, could not delete object in use
2. Admin: Can we stop object from being used?  No.
3. Admin: override to push through Deploy #3721 and delete the object in use

So not a blanket deletion authorization for a particular role (although that would be good too, for admins) but a one-time exception for specific operations.  The key is being able to execute on that *without* changing the source code of the CI/CD system.

Reviewer 12: Also, if you keep it binary, and for specific operations, then the whole "graph problem" doesn't occur.  Either the operation has authorization to ignore liens, or it doesn't.

Author: > Question for breakage: is there some way that leins can be overridden for automated tools without needing to change the code of the tools? I'm thinking administrator action, but done outside what may be a third-party tool. If the answer is yes, then there's no compatibility problem.

As an advisory feature, any users who can update an object can force delete the object by (1) patching the object to remove all the liens on the object, and then (2) deleting the object, in most cases. Some users may think that this extra step to remove liens break the existing tool, but at least they can continue using their tools without changing the tools' codes or behaviors. I hope the answer to your question can be "yes" by this behavior.

Note that the force delete option here is for admins to avoid a special case where a kind of controllers, maybe not a human, watches the object and keep setting liens on it. Admins may eventually win the race to remove the liens before deletion, but it would be better to provide a way to avoid such situations. Also, allowing only certain users to specify this option is to avoid users or tools from always specifying this option, which makes this entire feature useless. 

In addition, for consuming controllers to avoid such races, they can be implemented to only add liens in certain conditions, like an object has a certain annotation, as described [here](https://github.com/kubernetes/enhancements/pull/2640/files#diff-7c2b9b5fce93fb956c6589eaefc44491414766ee66821cd4ddd4e4294449ab4bR186-R189). Such mechanism can be either opt-in or opt-out, but should be able to be changed later after creating objects. By using this kind of features on the consuming controllers side, users or tools will be able to delete objects without force delete and race.

- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

This KEP proposes a generic feature to protect objects from deletion while it is marked as in-use.

> ## Motivation

Reviewer 13: Can you also add a quick summary of how this is different from finalizers and what are the problems that this is able to solve that finalizers can't?

Author: Thank you for your comment.

> what are the problems that this is able to solve that finalizers can't?

Yes, it needs to be clear in this section. Added.

> how this is different from finalizers

It is described in "Design Details" as "(Just to be clear, the difference between `Liens` and `Finalizers` is that `Liens` blocks the deletion request itself, while `Finalizers` blocks the deletion to be completed)."

If anything need to be added here or need to be moved to "Motivation", please point out.

Reviewer 13: Thanks this is helpful!

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

## Alternatives

<!--
What other approaches did you consider, and why did you rule them out? These do
not need to be as detailed as the proposal, but should include enough
information to express the idea and why it was not acceptable.
-->
- Each controller for protecting objects implements its own logic. However, it requires much implementations for the same logic and potentially provides inconsistent user interfaces, like many different finalizers and different ways to opt-out, to users.

> - Implement similar mechanism by using finalizers or admission webhook to block deletion

Reviewer 14: What are the tradeoffs between Liens and admission webhooks?  

It seems like there are many more examples of admission webhooks than there are examples of finalizer use?  Should we prefer a pattern which is more widely used?

Reviewer: Admission webhooks are not very comparable in this case. Yes, you could make one that did the trick, but the real solution to this problem is a Schelling point. That is, the problem isn't "keep my object from being deleted" it's "let parties who know nothing of each other communicate the need to preserve an object." In that sense, whatever is merged upstream will win as long as it's not grossly ill-suited. And it is easier for us to add some apiserver logic than to operate a webhook.

Potential use cases are:
- Protect `PersistentVolume` while being used by `PersistentVolumeClaim` (Replace the background logic for pv-protection)
- Protect `PersistentVolumeClaim` while being used by `Pod` (Replace the background logic for pvc-protection)
- Protect `Secret` while being used by `Pod`, `PersistentVolume`, `VolumeSnapshotContent` (Secret protection above)
- Protect `ConfigMap` while being used by `Pod`
- Protect `StorageClass` while being used by `PersistentVolume`
- Protect `VolumeSnapshot` while being used by `VolumeSnapshotContent`
- Protect `VolumeSnapshotClass` while being used by `VolumeSnapshotContent`
- Protect `User`, `Group`, `ServiceAccount` and `Role` from `RoleBinding`
- Protect `User`, `Group`, `ServiceAccount` and `ClusterRole` from `ClusterRoleBinding`
- Protect resources that users marked as in-use

### Goals

- Provide a generic mechanism that prevents a resource from being deleted when it shouldn't be deleted

### Non-Goals

- Provide a specific mechanism to decide which particular objects should be prevented from being deleted when other particular objects exist

> - Provide a generic mechanism that prevents a resource from being updated when it shouldn't be updated

Reviewer 3: - Implement any of the above potential use cases

### User Stories (Optional)

#### Story 1

Another controller to protect Secret from deletion watches all Secrets and their potential user objects, like `Pod`, `PersistentVolumes`, and `VolumeSnapshotContent`. Once it find the Secret is used by one of the objects, it updates the Secret's `Liens` and ask this mechanism to block deletion request for the secret.

#### Story 2

A user sets `Liens` to an object to be protected by this mechanism. For example, when a user knows that `Namespace` X shouldn't be erroneously deleted while `Deployment` Y exists for his/her particular use case, the user can set such a protection.

<!--
### Notes/Constraints/Caveats (Optional)
-->

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

Most of the design ideas come from [here](https://github.com/kubernetes/kubernetes/issues/10179).

> Users or controllers can add a string to `Liens` to ask to protect a resource.

Reviewer 3: Should we consider a subresource for this, so we can give tighter RBAC?

Author: I will check other KEPs on what should be needed to define this as a subresource and update the KEP (may be along with the final version of lien field).

Reviewer 3: that could be something we evolve to for beta

Author: Added as a beta graduation criteria.

Author: Added as a beta graduation criteria.

Reviewer 6: I think this should be a requirement for alpha so that we can get feedback about the actual API shape and so that clients will not have a guaranteed forced migration.  I've mentioned similar here: https://github.com/kubernetes/enhancements/pull/2840#issuecomment-1024238332

`Liens` is defined as key-value pairs of string and list of struct, where key is used to identify users or controllers and value is used to describe the purpose of the use.
Multiple users or controllers can add their `Liens` for their own purpose, and they can remove their own `Liens` when it is no longer needed.
Deletion requests of the resource are blocked until the last `Liens` for the resource is removed
(Just to be clear, the difference between `Liens` and `Finalizers` is that `Liens` blocks the deletion request itself, while `Finalizers` blocks the deletion to be completed).

To make multiple users and controllers share the `Liens` field properly, they should include their identity to the `Liens` string and they should only delete their own `Liens` strings.

A PoC implementation for in-use protection can be found, [here](https://github.com/mkimuram/kubernetes/commits/lien).
Also, how it can be consumed by Secret protection controller can be found, [here](https://github.com/mkimuram/secret-protection/tree/lien).

### API Changes

`Liens` is added to `ObjectMeta`.

```go
Liens map[string][]Lien
```

Where `Lien` is defined as follows:

```go
type Lien struct {

> 	LienID string

Reviewer 3: I'm not conviced we need all this - can you justify it all?

Reviewer 2: +1

Also - my comment below about growing the object indefinitely is somewhat related to it.

Reviewer 3: yes - it seems hard to make this list bounded.  e.g. if 1000 pods depend on a secret, will we allow 1000 liens?  How about 10,000?  Those numbers are not unrealistic.

Author: As for the secret protection use case, all PVCs that are created from the same StorageClass may reference to the same single Secret, if the StorageClass reference to a single specific Secret. If the number of such PVCs become 1,000, it easily happens.

Author: Changed back to slice-of-strings Liens and restrict it to per-controller/per-user.

When deletion of the namespace A is requested, the deletion isn't completed until the object with lien is deleted.

#### Block Adding Additional `Liens` while Deleting

`Liens` shouldn't be added after `DeletionTimestamp` is non-nil, which means pre-deletion processes for finalizers are being handled.
Adding `Liens` while deleting itself won't do any harm for the cluster, because the resource will be deleted without any additional deletion requests that will be blocked by `Liens`.
However, users may think that the successful addition of the `Liens` means that the resource won't be deleted until the `Liens` are deleted, which isn't true.
To avoid such a misunderstanding from users, a request for adding additional `Liens` to a resource with non-nil `DeletionTimestamp` should be blocked by api-server.

#### Per Controller `Liens` vs Per Reason `Liens`

The same controllers may need to add `Liens` to a certain resource for multiple reasons, like multiple resources are using it.
In such a case, if `Liens` are added per controller basis, such a controller needs to be sure that no other resources are actually using the resource, when removing its `Liens` from the resource.

> On the other hand, if `Liens` are added per reason basis, such a controller can remove the `Lien` for the resource, without considering other resources.

Reviewer 3: So (again, tell me if I am wrong) you're saying it's not like finalizer which is just saying "wait", but it really does model a dependency graph?

Reviewer 3: maybe pseudocode for such a controller would help?

Reviewer 3: This *is* effectively a dep graph.  See above comments about the size of this list.  a `[]string` API here means that we can put an effective limit like "only 256 liens" allowed and add monitoring and see if that is actually a problem limit.  once you enumerate the lien-holders, the limit MUST be larger, which puts the whole system at risk.

Author: > I'm personally against having effectively a lien for every single "referencee".

I agree with you that adding a big list isn't good from a scalability view point.

On the other hand, adding a list of used resources will give users a better understanding on which resources are actually using it. In other words, it will let users know which other resources should be deleted to delete the resource safely. It will be a big benefit from a UX view point.

In addition, from the lien consumers' implementation view point, checking all the potential using resources before deleting a lien will be a time consuming operation. For example, in secret protection case, the controller needs to loop for all the secret to check whether any Pods or PVCs are using the secret, in order to delete the per controller lien (Per object lien will allow controllers just to check if the resource that is referenced from per object lien is still using the Secret).

Actaully, pv-protection and pvc-protection are using the similar approach to the per controller lien case above by using finalizer. However, it just checks whether any Pods or PVCs are using the secret after controller observes non-nil DeletionTimeStamp on the secret, or only after users request to delete it. Similar approach won't be used for lien, because the lien blocks deletion request itself and there will be no room to add such a check for controllers. Therefore, controllers need to keep the lien up-to-date all the time. (For secret-protection to take the similar approach to finalizer, it needs to be implemented as a validation webhook for deletion that checks whether any Pods or PVCs are using the secret, instead of using lien. It means that lien becomes useless for this kind of use cases).

Author: > I will consider whether we can set an effective limit or there will be other good approaches.

I'm considering if there is a way that consumers can handle per controller lien efficiently.

@wojtek-t 's comment [here](https://github.com/kubernetes/enhancements/pull/2640#discussion_r703553386) let me think whether current secret-protection-controller logic with per object lien is really simple (Whether creating an in-memory dep-graph in such controllers, like the garbage collector does, might be better, and so on. Per object lien may be still better from the UX view point above, though).

Any feedback and ideas are welcome, as we are running out of time.

Reviewer 2: > In addition, from the lien consumers' implementation view point, checking all the potential using resources before deleting a lien will be a time consuming operation. 

I kind of disagree.
A controller may keep the in-memory mapping from secret to list of objects that are referencing it.

Let's for a moment assume that only Pods can reference it - this trivially generalizes for other objects.
1. Updating that mapping is simple - whenever some Pod is created/updated/delete - we simply update the mapping for all secrets it is referencing.
[FWIW - we use a similar approach in secret-manager in kubelet.]
2. Whenever we change the number of references from 0 to more than 0 or vice versa, we update the lien of that secret.

So that's actually fairly efficient implementation that we can have.

BTW - another advantage of that is that we significantly reduce the number of API calls (as we don't have to update the secret on every new reference to it) - which is another huge advantage of this approach.

Author: It seems that there will be a way to efficiently update lien by consumer side, so I will update the KEP to only allow per controller basis. Thank you for your review and sharing the ideas in detail. @wojtek-t

`Liens` allows both per controller `Liens` and per reason `Liens`, and reason may exist per object.
Only requirements are:
  - Use key as identity of controllers or users that set the lien, and only add/remove its own list of liens
  - `lienID` is the only mandatory field in list of liens, and it needs to be unique across the lien for a key

Examples of the use of lien are as follows:

Let's assume that a secret protection controller would like to protect Secret `my-secret` in `ns1` when
both PersistentVolume `my-pv1` and Pod `my-pod1` in namespace `ns2` are using it.

- Per Controller `Liens`:

```yaml
apiVersion: v1
kind: Secret
metadata:
  namespace: ns1
  name: my-secret
  liens:
    k8s.io/secret-protection-controller:
      - lienID: used
```

- Per Reason `Liens`:

```yaml
apiVersion: v1
kind: Secret
metadata:
  namespace: ns1
  name: my-secret
  liens:
    k8s.io/secret-protection-controller:
      - lienID: used by PV(s)
      - lienID: used by Pod(s)
```

- Per Object `Liens`:

```yaml
apiVersion: v1
kind: Secret
metadata:
  namespace: ns1
  name: my-secret
  liens:
    k8s.io/secret-protection-controller:

>       - lienID: v1/PersistentVolume//my-pv1

Reviewer 2: Wearing my scalability hat, I would like to ensure that this won't be growing the object too far.

Reviewer 2: I commented a bit more on that above.

you need any help or guidance.
-->

### Feature Enablement and Rollback

<!--
This section must be completed when targeting alpha to a release.
-->

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: InUseProtection
  - Components depending on the feature gate: kube-apiserver
- [ ] Other
  - Describe the mechanism:
  - Will enabling / disabling the feature require downtime of the control
    plane?
  - Will enabling / disabling the feature require downtime or reprovisioning
    of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).

###### Does enabling the feature change any default behavior?

Deletion requests for an object are blocked while its `Liens` field is non-empty.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes, by disabling the feature gates.

###### What happens if we reenable the feature if it was previously rolled back?

Deletion requests for an object are blocked while its `Liens` field is non-empty, again.

###### Are there any tests for feature enablement/disablement?

> Yes, unit tests cover scenarios where the feature is disabled or enabled.

Reviewer 2: I'm assuming they aren't yet exist (as the feature doesn't yet exist) - they will be added prior alpha release?

Reviewer 2: So let's make it explicit for now, sth like:

"Tests covering feature enablement/disablement will be added prior Alpha release"

[or sth along those lines]

Author: Added.

Author: Added.

Reviewer 2: nit: let's remove this line then - it's kind of contradictory with the one below :)

-->

###### Does this feature depend on any specific services running in the cluster?

<!--
Think about both cluster-level services (e.g. metrics-server) as well
as node-level agents (e.g. specific version of CRI). Focus on external or
optional services that are needed. For example, if this feature depends on
a cloud provider API, or upon an external software-defined storage or network
control plane.

For each of these, fill in the following—thinking about running existing user workloads
and creating new ones, as well as about cluster-level services (e.g. DNS):
  - [Dependency name]
    - Usage description:
      - Impact of its outage on the feature:
      - Impact of its degraded performance or high-error rates on the feature:
-->

### Scalability

<!--
For alpha, this section is encouraged: reviewers should consider these questions
and attempt to answer them.

For beta, this section is required: reviewers must answer these questions.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.
-->

###### Will enabling / using this feature result in any new API calls?

> No.

Reviewer 2: Not directly - but they provide mechanisms that means that deletion may need more requests.

This is fine - I'm overall supportve for the idea even wearing my scalability hat, but let's maybe make it a bit more explicit.

Potential use cases are:
- To guarantee proper order of deletion:
  - Protect `PersistentVolume` while being used by `PersistentVolumeClaim` (Replace the background logic for pv-protection)
  - Protect `PersistentVolumeClaim` while being used by `Pod` (Replace the background logic for pvc-protection)
  - Protect `Secret` while being used by `Pod`, `PersistentVolume`, `VolumeSnapshotContent` (Secret protection above)
  - Protect `ConfigMap` while being used by `Pod`
  - Protect `StorageClass` while being used by `PersistentVolume`
  - Protect `VolumeSnapshot` while being used by `VolumeSnapshotContent`
  - Protect `VolumeSnapshotClass` while being used by `VolumeSnapshotContent`
  - Protect `User`, `Group`, `ServiceAccount` and `Role` from `RoleBinding`
  - Protect `User`, `Group`, `ServiceAccount` and `ClusterRole` from `ClusterRoleBinding`
  - Protect dependent resources while owner resources aren't request to be deleted
- To avoid accidental deletion of important resources:
  - Protect resources that controllers or users marked as important

### Goals

> - Provide a generic mechanism that prevents a resource from being deleted when it shouldn't be deleted

Reviewer 15: Does this KEP also cover to provide this feature for CRDs?

### User Stories (Optional)

#### Story 1

Another controller to protect Secret from deletion watches all Secrets and their potential user objects, like `Pod`, `PersistentVolumes`, and `VolumeSnapshotContent`. Once it find the Secret is used by one of the objects, it updates the Secret's `Liens` and ask this mechanism to block deletion request for the secret.

#### Story 2

A user knows this is an important object, and that terraform likes to delete and recreate objects, so the user sets a lien so that an accidental deletion will fail.

<!--
### Notes/Constraints/Caveats (Optional)
-->

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

Most of the design ideas come from [here](https://github.com/kubernetes/kubernetes/issues/10179).
Users or controllers can add a string to `Liens` to ask to protect a resource.

> A newly introduced validation in api-server will reject deletion requests for resources with non-empty `Liens`.

Reviewer 3: what error code?

"402 Payment Required" seems apt :)

"423 Locked" seems OK

`Liens` is defined as a slice of strings, like `Finalizers`.
The strings need to be namespaced keys.
Multiple users or controllers can add their `Liens` for their own purpose, and they can remove their own `Liens` when it is no longer needed.
`Liens` should be added per controller or per user basis.
Deletion requests of the resource are blocked until the last `Liens` for the resource is removed
(Just to be clear, the difference between `Liens` and `Finalizers` is that `Liens` blocks the deletion request itself, while `Finalizers` blocks the deletion to be completed).

A PoC implementation for in-use protection can be found, [here](https://github.com/mkimuram/kubernetes/commits/lien).
Also, how it can be consumed by Secret protection controller can be found, [here](https://github.com/mkimuram/secret-protection/tree/lien).

### API Changes

`Liens` is added to `ObjectMeta`.

```go
Liens []string
```

Validation criteria of the field are as follows:
- Keys must be namespaced like "kubernetes.io/secret-protection" and "foo/bar".

> - Maximum length of each key is 64 characters

Reviewer 3: 64 may be tight - I think we allow 253 for most label-like strings

Reviewer: +1

> Maximum number of keys is estimated around 4-8, but it is not set as a hard limit, because setting a hard limit will make the behavior inconsistent (which key can't be added depends on the order of addition).

Reviewer 3: I think we should add a limit of something unimaginable, like 32, so people don't use this with the wrong cardinality (e.g. add one lien per pod).  We can always bump 32 to 128 if someone can show me why that isn't ludicrous

Reviewer: +1, limits always have the property that a different actor might lose out on a retry, that's not a reason not to have them.

- Background cascading deletion: Deployment A is deleted immediately

#### Namespace Deletion

It may also not be clear how [namespace deletion](https://kubernetes.io/docs/tasks/administer-cluster/namespaces/#deleting-a-namespace) behaves with lien.
Therefore, this section describes the behavior.

In summary, lien will block the namespace deletion, but would not block deletion of any resources in the namespace individually.

If a lien is set to a namespace, only a deletion request to the namespace is blocked.
Users can still request to delete each object in the namespace directly, but they can't request to delete it through the namespace deletion.

If a lien is set to one of objects in a namespace, only a deletion request to the object is blocked until all other resources without liens are deleted.
This means a resource with lien will eventually be deleted.
Users can still request to delete its namespace or other objects in the namespace.

When deletion of the namespace is requested, the resource deletion in the deleted namespace is done by namespace lifecycle controller in below two steps:
1. Delete all resources (as a result, only resources without lien are deleted)

> 2. Force delete all resources (as a result, all resources are deleted regardless of lien)

Reviewer 3: So putting a lien on a pod and then deleting the namespace will ultimately override the lien?  Is that true with finalizers, too?  I think the behavior should be the same

Reviewer 2: It's not true for finalizers - finalizers has to be explicitly deleted.

I personally agree with the sentiment to unify their behavior, but IIRC it was @deads2k concern that adding a new mechanism that could break your automation (if e.g. namespace deletion doesn't delete the objects even though it previous was because now someone added a lien). I don't think it fully persuades me, but it's definitely a valid one.

Reviewer 3: Is this the main unresolved issue here?

Reviewer 10: We could write up the concerns and agree not to move to beta until there's evidence that these are covered.

I'm thinking that there will be 2 cases:
1. same-namespace liens
1. liens between a namespaced object and something not in that namespace

the second case seems like a good example of something that might block namespace deletion indefinitely; for the first case, you'd hope to see the controllers able to make progress and perhaps even to reflect that progress back via Events / status / whatever.

Therefore, a command line option is better suited to this feature.

#### Subresource

There is a use case to disallow a controller to write to its CRD except for the status subresource, but allow it to update finalizers/liens.
To allow this use case, it is required to provide a feature to set a separate RBAC to update finalizers/liens.
On the other hand, finalizers field is currently not subresourced.
Therefore, it needs to provide this feature in a compatible way for finalizers, and it would be better to implement liens field in the same way to finalizers.

To solve this issue, there is [an idea](https://github.com/kubernetes/enhancements/pull/2840#discussion_r723464012) to add a new nob for just patching only a specific field via an endpoint of a subresource, without modifying the field itself.
In this way, setting a separate RBAC will be acheived later without making liens field a subresource.
Therefore, liens field isn't defined as a subresource and providing a separate RBAC for liens is done in a separate KEP/PR along with finalizers.

#### Unresolved issues

- [ ] Choice of an advisory feature

> - [ ] Subresource (Is the conclusion that it is needed but will be added later OK?)

Author: This is about https://github.com/kubernetes/enhancements/pull/2840#discussion_r715140512 and Subresource section in this KEP. Please give feedback so that we can close this item.

> - [ ] Compatibility to existing deletion flows for existing tools (Is current behavior OK and current description enough?)

Author: There was a good discussion on this topic [here](https://github.com/kubernetes/enhancements/pull/2840#discussion_r715608153) and I believe that the conclusion is that current behavior is OK. If there are any other concern remaining, please respond to there. Also, if any of the point discussed there need to be described in the KEP, please point out.

### User Stories (Optional)

#### Story 1

Another controller to protect Secret from deletion watches all Secrets and their potential user objects, like `Pod`, `PersistentVolumes`, and `VolumeSnapshotContent`. Once it find the Secret is used by one of the objects, it updates the Secret's `Liens` and ask this mechanism to block deletion request for the secret.

#### Story 2

A user knows this is an important object, and that terraform likes to delete and recreate objects, so the user sets a lien so that an accidental deletion will fail.

<!--
### Notes/Constraints/Caveats (Optional)
-->

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

Most of the design ideas come from [here](https://github.com/kubernetes/kubernetes/issues/10179).
Users or controllers can add a string to `Liens` to ask to protect a resource.

> A newly introduced validation in api-server will reject deletion requests for resources with non-empty `Liens` and return ["409 Conflict" error code](https://datatracker.ietf.org/doc/html/rfc2616#section-10.4.10).

Reviewer 16: This puts the onus on clients to keep retrying if they want something to be deleted, right?

I prefer the declarative, eventually consistent nature of the current PVC protection implementation where a client can declare that it wants a thing to be deleted, and once it's safe to delete the thing, it is. We've built similar machinery to this in Crossplane and I believe it works pretty well.

For our use case "eventually consistent deletes" has nice symmetry with other operations. Most Crossplane resources correspond to some external system like a cloud provider. Let's say I want to create a GCP Subnet with Crossplane, but doing so requires the VPC it lives in to exist first. We don't reject the creation of the Subnet object in the API server because the VPC doesn't exist. Instead we accept the desired state (that a Subnet should exist) and our controller keeps trying to create it. Presumably it will eventually succeed when when the VPC exists.

Is there some way to address the requirement that deletes "don't start" (in the controller-reconciling-the-delete) sense without out-right rejecting the desire to delete a resource by an entity that has permission to do so?

Reviewer 3: > where a client can declare that it wants a thing to be deleted, and once it's safe to delete the thing, it is

This is a very different semantic than "I want to delete this" -> "no".  The only way I know to achieve what you describe is with a finalizer, but this is not the right semantic.  At that point, the object in question is VISIBLY being deleted and async controllers are doing their own cleanup.  This is way to signal "begin the end-of-life process", not a way to delay it it.

Is there a different mechanism you are thinking of?

Reviewer 16: > This is a very different semantic than "I want to delete this" -> "no".

Right. This KEP appears to propose that all forms of "in-use protection" use "I want to delete this" -> "no" semantics. Those semantics appear to be different from how the contemporary PV and PVC implementations of in-use protection work, which are more like "I want to delete this" -> "okay, eventually". For Crossplane's use-case, I prefer the semantics of the contemporary implementations work to what is proposed by this KEP.

My understanding is that PV and PVC in-use protection are more like "I want to delete this" -> "okay, eventually, when not in-use". The deleted objects transition into a `Terminating` state and go away once they're no longer in-use. We have the same "okay, eventually" semantics for Crossplane's `ProviderConfig` in-use protection, and are looking to generalize the pattern.

> Is there a different mechanism you are thinking of?

I'm not sure what you're asking. Can you elaborate?

To be clear - I think I understand the issues with finalizers and agree that they're not a good solution. I also don't have a good solution in mind. 🙂 What I would like is that _somehow_:

* If a client asks for an object be deleted
* The object can't be deleted _yet_ because doing so would break some other, dependent object
* The delete (i.e. desired state) be accepted and reconciled when possible

I don't think my preferred semantics necessarily make sense for all forms of deletion protection. The "Motivation" section of this KEP states that it's attempting to solve for two use cases:

1. Guaranteeing proper order of deletion (i.e. the PV and PVC use case, and Crossplane's)
2. Avoiding accidental deletion of important resources

I feel that "I want to delete this" -> "no" semantics makes sense for the second use case, but not the first.

Reviewer 15: In KubeVirt we have a use-case with same or similar requirements:
1. We are running our processes in the original pod
2. There is an eviction request for this pod
3. Upon this request we want to transfer some state of the processes to a new pod
4. Upon the completed state transfer, the original pod is now free to be terminated

Important aspects are
a. Application can observe eviction request
b. Pod stays running until the state is transfered
c. caller of eviction is aware that the pod is preparing a shut-down

Reviewer 3: @fabiand We could model this a few ways, but evictions are generally "urgent", right?  It's not clear to me how that should intersect with any sort of deletion delay.

Reviewer 16: @thockin bringing your [related comment](https://github.com/kubernetes/enhancements/pull/2840#issuecomment-1585307886) back into this thread: 😄 

> To do what you want, we need a new lifecycle which includes some indefinite pre-delete state.

That may be the case, but what we want does seem in-scope for the goals defined by this design - i.e. guaranteeing proper order of deletion.

Take this manifest:

<details>
<summary> pod-with-volume.yaml </summary>

```yaml
---
apiVersion: v1
kind: Pod
metadata:
  name: task-pv-pod
spec:
  volumes:
    - name: task-pv-storage
      persistentVolumeClaim:
        claimName: task-pv-claim
  containers:
    - name: task-pv-container
      image: nginx
      ports:
        - containerPort: 80
          name: "http-server"
      volumeMounts:
        - mountPath: "/usr/share/nginx/html"
          name: task-pv-storage
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: task-pv-claim
spec:
  storageClassName: manual
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 3Gi
---
apiVersion: v1
kind: PersistentVolume
metadata:
  name: task-pv-volume
  labels:
    type: local
spec:
  storageClassName: manual
  capacity:
    storage: 10Gi
  accessModes:
    - ReadWriteOnce
  hostPath:
    path: "/mnt/data"
```

</details>

Today I can `kubectl apply -f pod-with-volume.yaml` and despite there being dependencies between the resources, the create just works and the controllers eventually arrive at my desired state - that they're created. I can `kubectl delete -f pod-with-volume.yaml` too. My desired state (that all this stuff be deleted) is immediately accepted, and eventually it's all gone.

I believe under this proposal the `kubectl delete -f pod-with-volume.yaml` UX would change. The `Pod` would delete immediately but the `PersistentVolume` and `PersistentVolumeClaim` would fail with a 409 Conflict, and I'd basically have to retry a couple of times for everything to actually get deleted.

This does technically achieve the goal of guaranteeing proper order of deletion - it just seems like a pretty poor UX for doing so.

Reviewer 15: @thockin I was not ware that evictions had the notion of urgency.
Evictions respect PDBs, evictions are less hard than DELETE. Thus isn't eviction rather the less urgent approach and delete the real urgent way of getting rid of workloads?
fwiw `kubectl drain --force --grace-period=0` is urgent.

Thus to me a delay to an eviction is similar to [systemd inhibitors](https://www.freedesktop.org/software/systemd/man/systemd-inhibit.html) - a way to allow workloads to shut down gracefully. or even fail. And that's the difference: Fail gracefully, give admin/owner time to eeact to this. Before an admin jumps in with `--force --grace-period=0`

Users can still request to delete Replicaset B and Pod C directly, but they can't request to delete them through Deployment A.

If a lien is set to Replicaset B, only a deletion request to Replicaset B is blocked.
Users can still request to delete Deployment A and Pod C directly.
When deletion of the Deployment A is requested, whether it completes immediately or not depends on the cascading policy and Replicaset B's `blockOwnerDeletion`.
The behaviors are as follows:

- Foreground cascading deletion:
  - `blockOwnerDeletion=true`: Deletion of Deployment A isn't completed until Replicaset B is deleted
  - otherwise: Deployment A is deleted immediately
- Background cascading deletion: Deployment A is deleted immediately

#### Namespace Deletion

It may also not be clear how [namespace deletion](https://kubernetes.io/docs/tasks/administer-cluster/namespaces/#deleting-a-namespace) behaves with lien.
Therefore, this section describes the behavior.

If a lien is set to a namespace, only a deletion request to the namespace is blocked.
Users can still request to delete each object in the namespace directly, but they can't request to delete it through the namespace deletion.

> If a lien is set to some of objects in a namespace, a deletion request to the namespace isn't blocked.

Reviewer 3: I want to register my disagreement with this, but I yield to api-machinery folks.

Author: @deads2k @lavalamp Could you review this KEP again?

Reviewer 17: Since this mechanism is intended for protecting against mistakes, it's not just a question "who has the power to", this is meant to prevent myself from mis-using my power. If I have my very important database that should absolutely not be deleted in a namespace, I don't want to be able to remove it accidentally by removing the namespace.

EDIT: Especially now that I see that there is a `IgnoreLiens` option available to force delete.

To solve this issue, there is [another KEP](https://github.com/kubernetes/enhancements/issues/3285) to add a new knob for just patching only a specific field via an endpoint of a subresource, without modifying the field itself.
In this way, a feature to set a separate RBAC can be implemented later without making liens field itself a subresource.

Subresourcing `Liens` is a beta blocker and won't be included in the initial implementation.

#### Unresolved issues

- Decision: When you delete a namespace and an object in that namespace has a lien:
  - Does the namespace and object get deleted anyway?
    - [x] Yes.
  - Does the namespace deletion get blocked (i.e., the delete request fails) if any object in the namespace has a lien?
    - [x] No.
  - Does the other objects in the namespace get deleted, but that not object, preventing the namespace deletion from completing?
    - [x] No. If you delete the namespace, the content will be removed.
  - Does the namespace deletion not proceed until all objects in the namespace are free of liens?
    - [ ] No. If you delete the namespace, the content will be removed. Namespace lifecycle cleanup will not honor liens.

>     - [x] No. If you delete the namespace, the content will be removed. Namespace lifecycle cleanup will honor liens only to guarantee the deletion order inside a namespace (guarantee of the deletion order is a beta blocker).

Reviewer 3: I'm not sure deletion order is important UNLESS we can ultimately leave the deletion pending

Author: I also prefer to make deletion of resources with lien pending as an initial implementation (in this case, how cluster admin force the deletion of namespace will be a beta blocker), but either way is fine to me as long as the final goal is the same.

Reviewer 6: > I'm not sure deletion order is important UNLESS we can ultimately leave the deletion pending

I think eventually deleting is important, but I can also imagine putting a timeframe on it (an hour?).  the namespace lifecycle controller could try to delete everything every few minutes and eventually, by blind luck, it will eventually make enough progress if an ordering exists.  After the timeout, the namespacelifecycle controller can override what remains.

The code that relies on finalizers today still needs to keep those finalizers in order to guarantee the cleanup.  Liens just make it less likely to have deletion in a "bad" order.

Reviewer 6: >I think that we still have choices:
>to make an hour time out happen unconditionally

unconditionally is better.

  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

> This KEP proposes a generic feature to protect objects from deletion while it is marked as in-use.

Reviewer: ```suggestion
This KEP proposes a generic feature to protect objects from deletion while they are marked as in-use.
```

## Motivation

Currently, "a generic mechanism that prevents a resource from being deleted when it is in-use" doesn't exist.
Instead, for such a use case, each controller, like [pv-protection](https://github.com/kubernetes/enhancements/issues/499) and [pvc-protection](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/postpone-pvc-deletion-if-used-in-a-pod.md), implements its own logic to protect used objects from being deleted while in-use.
These controllers use [Finalizers](https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/) to block deletion.
Finalizers block the _completion_ of a delete operation, but they do not prevent the deletion from starting.
Once started, a delete _will_ complete and can not be aborted.

> However, Finalizer may not be appropriate for some resource protection use cases, because it doesn't prevent other controllers from executing their pre-deletion actions.

Reviewer: ```suggestion
Finalizers may not be appropriate for some resource protection use cases, because they don't prevent other controllers from executing their pre-deletion actions.
```

The order of execution of pre-deletion actions across finalizers is not defined.
As a result, other controllers will execute their pre-deletion actions while a finalizer for protection still exists.

This KEP aims to provide a generic mechanism for protection that solves the above issues with Finalizers, because such use cases are also found for protecting secrets and configmaps in [KEP-2639](https://github.com/kubernetes/enhancements/pull/2640).

> Potential use cases are:

Reviewer 18: Based on these stated use cases, I don't think Liens is the right solution to this problem.

All of these are examples of object dependencies. Some of them are parent-child relationships, and others are sibling relationships. But they're all dependencies.

Here's some of my reasoning...

### Liens are hard to understand & debug
Also, my experience using Liens on other platforms (like GCP), I have always found them to be very hard to understand and debug for users. A dependency, on the other hand, is a well understood concept intuitively graspable by all technical users. A blocked action from a dependency can tell you exactly what dependencies block the action, which can be used to describe _why_ the aciton was blocked. A Lien on the other hand does not describe why, only that the action is prohibited. To learn why you have to dig around to figure out what made the Lien and ask them why.

### Liens are negative by connotation
Liens are something the IRS takes out on your house and banks take on your car. They're almost always interpreted as negative, even if they do facilitate lending and repayment. I've never had a user ask me to implement liens in a software. What they ask for is policies, constraints, and dependencies.

### Liens are an intermediate data format
Prevention of actions is easily implementable with a webhook. So enforcement of Liens is simple. The complex part is defining them. And if you define them as Liens, you're effectively defining an intermediary data format between intent and behavior. Why not just describe the intent?

### Finalization and Initialization parity
We already have a problem that the behavior from Finalizers are not replicated by Initializers. I'm loath to put another Finalization-like behavior into the API without thinking through the inverse Initialization behaviors that we might not be able to achieve if we choose the Finalization interface first. If you think about liens as "deletion ordering requirements" it becomes obvious that there are also "creation ordering requirements" in the opposite order. 

### Apply ordering
If you try to implement creation and deletion ordering (as I have in [kpt dependency ordering](https://kpt.dev/book/06-deploying-packages/03-handling-dependencies)), it becomes apparent that you might also need  "apply ordering requirements" to apply updates in the right order across multiple resource objects. In fact, kubectl already does this when it applies a directory, by applying the files alphanumerically and top to bottom. This allows users to specify the order that objects get applied and deleted, when done so all together with a single kubectl command.

### Object References
Many of the uses cases here already have ways to reference the object they depend on (dependent references dependency), for example PersistentVolumeClaim `spec.storageClassName`. Sometimes, the reference is inverted (dependency references to dependent), for example PersistentVolumeClaim `spec.claimRef`, which is used for accounting. If we could standardize these types of references, we could define the dependency generically on the CRD as being a quality of the resource, and not something specific to each object. And of course, what we can do for CRDs in yaml, we can do in code for core objects and extensions.

So in summary, instead of Liens, I would rather pursue object references and dependencies. This way we can then implement not just deletion prevention, but also client-side create/apply/delete ordering. Granted, that's a more ambitious project, but it's also a better more holistic solution.

/cc @bgrant0607 @thockin @apelisse

Reviewer 3: Should we have generalized referential integrity in Kubernetes?  Maybe.  It would be a VERY different system and I just don't see a path from here to there in any realistic timeframe.

I mean, it's "just software".  We could do anything IN THEORY, but I don't see this happening any time soon (and I can think of a LOT of things that I'd spend time on first).

I hate the idea of calcification of systems, but I also hate not doing X because Y would be better, and then not doing Y.  If someone wants to explore how to build the full graph of resources into the apiserver (where each edge carries policy), and figure out how that changes the sytem overall (compatibly, of course), and how it scales at the API and storage layer, I am more than willing to help chart the course.  I have a lot of ideas and even more questions.

But I don't think we should derail useful incremental work in the mean time.

Reviewer: Apiserver has no control over the order of multi-object apply/creates. We took initializers out because it was too hard to make the entirety of the system deal with partially constructed objects, so I don't see this changing.

Additionally, we are almost certainly never going to have transactions.

We actually already have dependencies -- ownerReferences. But they are for auto-tear-down, not for preventing deletion.

I do agree with the point about liens being hard to debug, and it's worth thought about how we can make that understandable. I don't think liens should have to be inherently more hard to understand than a dependency, which can also be inscrutable (why does X depend on this random custom resource??). Additionally you can implement a dependency system using liens, if you want to.

API features are often "some assembly required" because that way you can solve more use cases with fewer mechanisms (moving complexity for specific use cases into plugins/controllers).

Reviewer 3: To be clear, I could imagine a world where:

* Pointer-like fields are tagged "this is a pointer (to a fixed resource or a dynamic one)"
* Pointer fields carry policy like
  - "inhibit deletion of the referent"; or
  - "delete me when the referent is deleted"; or
  - "inhibit deletion of me as long as the referent exists"; or
  - "fail creation if the referent does not exist"
* "The system" enforces those edges

What I don't know is, well, everything else.  How will it scale?  How will it impact API complexity?  How do I prevent errant referrers?  How do I force-override these?  what fun new consistency issues arise?  

It's just a LONG way from here.  It sounds like a fun project, though.

Reviewer 18: > I hate the idea of calcification of systems, but I also hate not doing X because Y would be better, and then not doing Y... But I don't think we should derail useful incremental work in the mean time.

In a world where reverse incompatible API changes are possible with major version bumps, I agree that preventing incremental change is bad. But I think the Kubernetes reverse compatible guarantees have calcified the system in such a way that making a bad API change (one that prevents future iteration by not planning for it) is worse than not making that change.

>  If someone wants to explore how to build the full graph of resources into the apiserver (where each edge carries policy), and figure out how that changes the sytem overall (compatibly, of course), and how it scales at the API and storage layer, I am more than willing to help chart the course. I have a lot of ideas and even more questions.

We're already built a system into kpt that builds a dependency graph and handles ordering of applies and deletes. It still needs some work before I would recommend it for general use (I plan to make it do parallel apply, instead of serial apply), but it's not particularly _hard_ to build. It's "just software" as you say.

The key constraint in what kpt does today is that it's limited to a single "package" or set of objects managed together. Managing incoming dependencies from any object on the cluster would have to be server-side, not client-side. But that's something that could be enforced by a webhook, even if the webhook can't perform the ordering of applies/deletes. A webhook would simply reject out-of-order requests with an error message that explains the reasoning. As long as a client can interpret the same dependencies, the client can perform ordering of operations for the set of objects it cares about.

> Apiserver has no control over the order of multi-object apply/creates. We took initializers out because it was too hard to make the entirety of the system deal with partially constructed objects, so I don't see this changing.

Agreed. I don't think the server can or should re-order apply operations. To do so would require batch operations and/or transactions, which, as you say, is unlikely.

However, it's entirely reasonable for the server to simply to enforcement to prevent unordered operations (deletion ordering & creation ordering). Then it's up to the client to handle ordering operations so that they satisfy those requirements.

Reviewer: All that stuff can be built via webhooks if the system offered the lien feature. It is tricky to get it right. I want people to experiment and build it. But it's very slow and hard to do so in-tree. Plugins don't have compatibility problems until people actually start using them, so they're much faster to develop.

I see liens as enabling that behavior, not as competing with it. (You can't build this today because there's no way to prevent a deletion.)

Reviewer: > webhooks can prevent a delete request, right?

Good point. In that case I'd like to see an out-of-tree prototype built before we do something in-tree?

Reviewer 18: A few other points:

### Self Liens
It's nonsensical for a property owner to take a lien on their own property. Liens are always taken (in real life) by a 3rd party who needs to legally prevent the owner from performing a sale or transfer. So it doesn't make sense to use a "lien" to allow an owner to prevent "accidental deletion" of their own objects. Liens ARE dependencies. So you might as well say "bank X requires payment Y before property Z is allowed to be sold. And then you can invert it into a construction policy: "city X requires permit Y before property Z can be built." They're both dependencies, but only one of them is a Lien.

## Deletion protection
Imagine the "don't delete this database" use case in GitOps. In interactive modes it might trigger an "are you sure?" prompt or require a "--force" flag. But in GitOps, neither of those are possible, because not only are the manifests declarative but the operations performed by the client are also declarative. The object either exists in the repo directory and is created/updated or it was removed and is deleted. How do you remove a Lien in this case? Is it really any better than just removing the object?

## Defining operation ordering
Say we add Liens. How does a user now apply their objects in the right order? With kubectl/kustomize, they have to alphabetize their files or arrange objects top to bottom in a file. This affects both create, update, and deletion ordering. With kpt you get some resource-based sorting for free (covers most of these uses cases) or you can manually define dependencies. But do you really want the user to need to manually order their applies? What if the server-side enforcement and apply/delete ordering both used the same interface? Wouldn't that be better?

Reviewer: In this system, typically the lienholder wouldn't be the object itself.

I think close to the whole point of liens is to break gitops workflows when they're about to clobber something important.

I don't see why liens affect creation order.

Reviewer 18: > All that stuff can be built via webhooks if the system offered the lien feature.

If liens only define deletion prevention, then no, the API won't be sufficient to also define creation and update ordering. A "lien" exclusively prevents deletion (aka sale or transfer). If we want it to also specify create & apply ordering, we can't call it a lien. Perhaps this is quibbling over naming, but I think the naming is important to the meaning the usage specifies.

Reviewer: I continue to not see why you think a lien has or should have an effect on creation or update ordering?

Reviewer 18: > I think close to the whole point of liens is to break gitops workflows when they're about to clobber something important.

There's two distinct use cases being described. One is that the object is in-use. The other is that the object is "important". Personally, I think if we define dependencies to satisfy the in-use cases (e.g. Deployment (backend) using a StatefulSet (Database)) then most of the other use cases evaporate, or can be defined in terms of a dependency.

> I don't see why liens affect creation order.

"Liens" themselves do not. However, if we define an interface for deletion order, we should also define an interface for creation order, otherwise we end up with the same asymmetry we have with Initializers/Finalizers. There isn't an inverse word for Lien (AFAIK) that we could add later to handle create ordering. But "Dependencies" works both ways.

Reviewer 18: > To be clear, I could imagine a world where:
> 
> * Pointer-like fields are tagged "this is a pointer (to a fixed resource or a dynamic one)"
> * Pointer fields carry policy like
>   
>   * "inhibit deletion of the referent"; or
>   * "delete me when the referent is deleted"; or
>   * "inhibit deletion of me as long as the referent exists"; or
>   * "fail creation if the referent does not exist"; or
>   * "fail updates that break this reference"; or
>   * ... ad infinitum
> * "The system" enforces those edges

Dependency references should be able to handle most of these:
- Disallow deletion of that object before this object is deleted (deletion ordering)
- Disallow creation of this object before that object exists (creation ordering)
- Disallow update of this object before that new object exists (update ordering)

The one left out is: 
- Delete me when the referent is deleted (garbage collection)
But that's already handled by resource ownership today.

> How will it scale? How will it impact API complexity? How do I prevent errant referrers? How do I force-override these? what fun new consistency issues arise?

Good questions. I don't have immediate answers. I didn't come with a full design spec'd out. Only a warning to not implement half the API in a way that prevents adding the rest later.

Reviewer: > , if we define an interface for deletion order, we should also define an interface for creation order, otherwise we end up with the same asymmetry we have with Initializers/Finalizers. There isn't an inverse word for Lien (AFAIK) that we could add later to handle create ordering. But "Dependencies" works both ways.

I'm not following this logic at all, I don't see that symmetry as existing *at all* in a world where clients do the create/update ordering and it can't be affected by the server. Even if it was a thing, I don't see how it would affect liens, IMO it is good that the name can only mean the one thing and we'd need a different name if we wanted it to do a different thing.

Reviewer: > I think if we define dependencies to satisfy the in-use cases (e.g. Deployment (backend) using a StatefulSet (Database)) then most of the other use cases evaporate, or can be defined in terms of a dependency.

I don't see it that way, I would prefer a very simple thing that works for all use cases than a complicated thing that we think might work for most use cases.

Reviewer 18: > I don't see it that way, I would prefer a very simple thing that works for all use cases than a complicated thing that we think might work for most use cases.

For most other projects I agree with you. But as I said earlier, I think Kubernetes' reverse compatibility guarantees makes it really hard to iterate on simple things, because once they're GA they can't be modified into a more complex thing without being reverse compatible. Naming is one big thing you can't take back, and naming informs the intent and usability here.

Reviewer 10: I saw this discussion and I thought of:

> [Cloud Native architectures] enable **loosely coupled** systems that are resilient, manageable, and observable

(my emphasis, quoting from the [cloud native definition](https://github.com/cncf/toc/blob/main/DEFINITION.md))

Reviewer: I talked with Karl to understand better his use cases and I promised to at least imagine how a dependency-based solution for this would look.

### User Stories (Optional)

#### Story 1

Another controller to protect Secret from deletion watches all Secrets and their potential user objects, like `Pod`, `PersistentVolumes`, and `VolumeSnapshotContent`. Once it find the Secret is used by one of the objects, it updates the Secret's `Liens` and ask this mechanism to block deletion request for the secret.

#### Story 2

A user knows this is an important object, and that terraform likes to delete and recreate objects, so the user sets a lien so that an accidental deletion will fail.

<!--
### Notes/Constraints/Caveats (Optional)
-->

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

Most of the design ideas come from [here](https://github.com/kubernetes/kubernetes/issues/10179).
Users or controllers can add a string to `Liens` to ask to protect a resource.

> A newly introduced validation in api-server will reject deletion requests for resources with non-empty `Liens` and return "423 Locked" error code.

Reviewer 10: Using a WebDAV status code feels wrong, especially given that 423 responses
> SHOULD contain an appropriate precondition or postcondition code, such as `lock-token-submitted` or `no-conflicting-lock`.

How about 400, 403, or 409?

Reviewer 3: I don't mind 423, but 409 seems OK: "The request could not be completed due to a conflict with the current state of the resource. This code is only allowed in situations where it is expected that the user might be able to resolve the conflict and resubmit the request."

Author: [Precondition/Postcondition XML Elements](https://datatracker.ietf.org/doc/html/rfc4918#section-16) for [423 Locked](https://datatracker.ietf.org/doc/html/rfc4918#section-11.3) looks a bit complex definition to use for this purpose. It is just a recommendation that is defined as "SHOULD" in RFC, so we will be able to ignore it, but we would be able to choose "409 Conflict" which doesn't have such recommendation and it would be good enough. So, I just changed to 409.

Reviewer 19: 409 is typically used for retry requests during race conditions, right? Would this prompt controllers to keep retrying the deletion at an incorrect backoff due to the assumption that it is just a concurrent write?

Reviewer 10: That's a valid concern @SOF3. 400 might be a better choice then, as the request is neither forbidden using access controls, conflicted (in the sense that another client's update has caused one) nor locked in the sense that WebDAV would mean it.

If we want to, we can include a message within the 400 response to explain why the request was rejected. Right?

`Liens` is defined as a slice of strings, like `Finalizers`.
The strings need to be namespaced keys.
Multiple users or controllers can add their `Liens` for their own purpose, and they can remove their own `Liens` when it is no longer needed.
`Liens` should be added per controller or per user basis.
Deletion requests of the resource are blocked until the last `Liens` for the resource is removed
(Just to be clear, the difference between `Liens` and `Finalizers` is that `Liens` blocks the deletion request itself, while `Finalizers` blocks the deletion to be completed).

A PoC implementation for in-use protection can be found, [here](https://github.com/mkimuram/kubernetes/commits/lien).
Also, how it can be consumed by Secret protection controller can be found, [here](https://github.com/mkimuram/secret-protection/tree/lien).

### API Changes

`Liens` is added to `ObjectMeta`.

```go
Liens []string
```

Validation criteria of the field are as follows:

> - Keys must be namespaced like "kubernetes.io/secret-protection" and "foo/bar".

Reviewer 10: No private liens? (Why not?)

Reviewer 3: Every time we have such indicator strings they fall into two camps:  `WellKnownThing` or `example.com/totally-extensible-thing`.  Occassionally, but rarely, we have a field that does both.

In this case I don't think we have any WellKnownThing liens, so I think we should error on the side of explicitness (I wish finalizers were also as explicit).

Reviewer 10: ```suggestion
 - Keys must be namespaced (example: `kubernetes.io/secret-protection`;  `foo.example/bar`).
```

These aren't actually _keys_, mind. Not in the key-value sense. For completeness we might want to clarify that duplicates are prohibited.

Reviewer 10: With labels and annotations, we let people use non-namespaced (private) keys. Unless we actively validate the lien names, I assume that folks will do that here anyway. What I'm wondering is why we're forbidding a private lien name?

If I run a cluster and, privately to that cluster, I want to use a lien that I have named `ingress-in-use-protection`, is that a bad thing? We should aim to explain _why_ before the beta, so that that reasoning can go in the docs.

Reviewer 3: For labels and annotations we leave non-prefixed names for end-users.

For resources we reserve non-prefixed names for well-known values.

We have never been consistent :(

In both cases it comes down to who uses them, how much, and which one do we optimize for.  End-users (developers, namespace users, non-admins) tend to be the ones spelling out labels and resource names, so we optimize for them writing less.  Most of the time they are writing their own labels and well-known resource names.

With liens, I don't expect end-users to use them almost ever.  It's not a burden to ask automated processes to use fully-qualified names, IMO.  That's my rationale.

`Liens` is added to `ObjectMeta`.

```go
Liens []string
```

Validation criteria of the field are as follows:

> - Keys must be namespaced (example: `kubernetes.io/secret-protection`;  `foo.example/bar`)

Reviewer 17: How is this validated? What characters are accepted? How is this different from finalizers? What are human beings supposed to use?

- Maximum length of each key is 253 characters
- Maximum number of keys is 32

### Other Design Considerations

#### Behavior with `ownerReference`

Lien itself only blocks a deletion request to an object that is added to.
However, it may not be clear how lien behaves when used with [`ownerReference`](https://kubernetes.io/docs/concepts/architecture/garbage-collection/#owners-dependents).
Therefore, this section describes the behavior.

In summary, lien will block cascading deletion, but would not block deletion of any dependent resources individually.

For example, let's assume that we have Deployment A which manages Replicaset B which manages Pod C.
In this situation, ownerRefernces are set from Pod C to Replicaset B and from Replicaset B to Deployment A.

If a lien is set to Deployment A, only a deletion request to Deployment A is blocked.
Users can still request to delete Replicaset B and Pod C directly, but they can't request to delete them through Deployment A.

If a lien is set to Replicaset B, only a deletion request to Replicaset B is blocked.
Users can still request to delete Deployment A and Pod C directly.

> When deletion of the Deployment A is requested, whether it completes immediately or not depends on the cascading policy and Replicaset B's `blockOwnerDeletion`.

Reviewer 20: Does this mean that the Lien is ignored and the resource is deleted when it's owner resource is deleted?  That would seem to imply that if you really need the Lien you should place it on the top-level resource and not on any of the lower-level resources.

Put another way, Liens are enforced by the API endpoint and NOT by the garbage collector.  So if an API Delete request is made for an object with a Lien, that gets enforced.  But if an object with a Lien is part of a set of ownerReferences and the top-level owner does not have a Lien, any Liens on the other resources in the graph are ignored.

Users can still request to delete its namespace or other objects in the namespace.

TODO: Update below description depending on the decision.

```
When deletion of the namespace is requested, the resource deletion in the deleted namespace is done by namespace lifecycle controller in below two steps:
1. Delete all resources (as a result, only resources without lien are deleted)
2. Force delete all resources (as a result, all resources are deleted regardless of lien)

It is to ensure proper order of deletion of resources as much as possible, even when they are force deleted through namespace deletion.
```

#### Block Adding Additional `Liens` while Deleting

`Liens` shouldn't be added after `DeletionTimestamp` is non-nil, which means pre-deletion processes for finalizers are being handled.
Adding `Liens` while deleting itself won't do any harm for the cluster, because the resource will be deleted without any additional deletion requests that will be blocked by `Liens`.
However, users may think that the successful addition of the `Liens` means that the resource won't be deleted until the `Liens` are deleted, which isn't true.

> To avoid such a misunderstanding from users, a request for adding additional `Liens` to a resource with non-nil `DeletionTimestamp` should be blocked by api-server.

Reviewer 10: ```suggestion
To avoid such a misunderstanding from users, the API server must block any request for adding additional `Liens` to a resource with non-nil `DeletionTimestamp`.
```

(changing a **should** to a **must**; we could defer implementing that 'til beta if it helps)

#### Format of `Liens`

> As a result of discussions, format of `Liens` is decided as a slice of strings and it needs to be used per controller(per user) basis.

Reviewer 10: ```suggestion
In Golang code to implement this behavior, the representation of `Liens` is a slice of strings. Duplicates are prohibited.
Liens need to be used per controller (per user / per dependency).
```

?

To mitigate the risk, a new `IgnoreLiens` API option in `DeleteOptions` to force delete a resource with liens will be added.

Specifying the option should only be allowed to a limited set of users and groups, therefore deletion request will be blocked in api-server, if invalid users request deletion with the option.

> Restricting the `IgnoreLiens` API option to specific users is beta blocker and won't be included in the initial implementation.

Reviewer 6: I think getting the permissions model right for alpha is important.  We don't want to needlessly break clients during an alpha to beta transition if we know this is coming.

Could this check actually be, "do you have the rights to modify the /liens subresource" since such permissions imply the ability to remove liens already and deleteoptions.ignoreLiens exists to eliminate the race.

Author: How about making DELETE permission for `{resource}/liens` subresource a requirement for `IgnoreLiens` API option (It may only be valid when DELETE permission isn't used for a normal use for subresourcing, but currently only GET, PATCH, and PUT verbs are planed to be used in [Subresource finalizers KEP](https://github.com/kubernetes/enhancements/issues/3285))?

Reviewer 10: > How about making DELETE permission for {resource}/liens subresource a requirement for IgnoreLiens API option..?

I like the overall idea here; I want to check that we would restrict at the Kubernetes verb level not HTTP. For example, you might use Server-Side Apply to remove all Liens (HTTP verb `PATCH`); RBAC verb might be `update` (repeatedly removing one lien) or `delete`, I'm not sure.

Reviewer 3: Is there a basic assumption that any user who can UPDATE/PATCH an object can clear liens?  If sho, we should say that.  Or do we need the ability to say "mere mortals cannot do that"?  E.g. some critical thing that must not be deleted until ACK'ed by a super-privileged actor?  I can't come up with any examples right now...

Assuming the above (any update/patch can clear liens), does the proposed check (delete `{resource}/liens`) pass for such users (e.g. I don't explicitly have permission to delete foo/liens but I do have permission to update foo) ?

    - Good scalability (Size of the field and number of API calls required)
  - Cons:
    - Bad UX (Users won't know which other objects are actually using the object)
    - Complex consuming controllers' implementation (Each controller needs to track the using objects)
- Per reason(Per object) `Liens`
  - Pros:
    - Good UX (Users can know which other objects are actually using the object)
    - Simple consuming controllers' implementation (Lien tracks the using objects)
  - Cons:
    - Not aligned with `Finalizers`

>     - Bad scalability (Size of the field and number of API calls required)

Reviewer 10: This could belong in the Alternatives section.

Author: I just deleted this section, because there is no discussion around it, recently. I can add it again if needed.

#### Race of removing/adding liens

A cluster-admin would be racing to update the object to remove the lien and delete the object before a namespace editor is able to place the lien back. As a result, a namespace admin or editor can create an object that a cluster-admin or namespace lifecycle controller cannot delete, which shouldn't be allowed.

> To mitigate the risk, a new `NotBlockOnLien` API option in `DeleteOptions` to force delete a resource with liens will be added.

Reviewer 3: How about `IgnoreLiens` as a name?

> TODO: Update below description depending on the decision.

Reviewer 10: Does this need addressing before merge?

> Subresourcing `Liens` is a beta blocker and won't be included in the initial implementation.

Reviewer 6: I think this needs to be resolved for alpha so that our clients will function properly across versions.  Similar comment here: https://github.com/kubernetes/enhancements/pull/2840#discussion_r898167946

Reviewer 3: My take: I'm assuming that mutating liens will be allowed through update/patch as well as a subresource (just like annotations and labels).  So a client written to use PATCH will still work, and when we add the subresource it will be *possible* to convert (but not *required*).

David - are you imagining otherwise?  If the ONLY way to mutate liens is thru the subresource (e.g. fail any update/patch which changes them -- which has a certain appeal) then I agree it is a requirement.

Reviewer 6: >  If the ONLY way to mutate liens is thru the subresource (e.g. fail any update/patch which changes them -- which has a certain appeal) 

I had been thinking of it in terms of "only the liens subresource modifies liens".  I can see a use-case for setting liens on creation, but for updates where the object has already existed without liens for some time, I don't see a reason why separate REST calls are particularly bad.

I can probably be persuaded that the CREATE and spec UPDATE are congruent.  But I see status update rules as more like what we built for CRs, where the *only* fields allowed to be updated are under status.

I don't see a way to build what I would consider a safe, general liens handler without a subresource and I think that sort of client handling is important to get right initially.

Reviewer 3: > I had been thinking of it in terms of "only the liens subresource modifies liens". 

Do we have any other examples?

Labels, annotations, and finalizers (all in metadata) can be updated by PATCH/PUT.

Scale (in spec) can be updated by PATCH/PUT *or* the subresource.

Status is the only one I know that is different.

I'm not against it - it's just different.

Also see https://github.com/kubernetes/enhancements/pull/2840#discussion_r900286510 - let's assume that the subresource is the ONLY way to update lines.  Alice is a namespace user - she can create Pods.  If a lien gets put on my pod - do we want Alice to be able to clear it?

Reviewer 17: > I had been thinking of it in terms of "only the liens subresource modifies liens"

that's not great for human beings adding the lien themselves, does it?

There was an alternative idea to mitigate the risk by making liens field a subresource to allow defining a separate RBAC.
By doing so, a cluster-admin will be able to set the RBAC for the liens to deny update by namespace admin or editor.
However, a command line option is chosen as a solution, because liens are introduced as an advisory feature to help prevent accidental deletion, not a hard locking feature which sets privileges around who can or cannot lock / delete.

As an advisory feature, anyone should be able to apply a lien, and anyone should be able to ignore them.
Therefore, a command line option is better suited to this feature.
```

#### Subresource

There is a use case to disallow a controller to write to its CRD except for the status subresource, but allow it to update finalizers/liens.
To allow this use case, it is required to provide a feature to set a separate RBAC to update finalizers/liens.
On the other hand, finalizers field is currently not subresourced.
Therefore, it needs to provide this feature in a compatible way for finalizers, and it would be better to implement liens field in the same way to finalizers.

> To solve this issue, there is [an idea](https://github.com/kubernetes/enhancements/pull/2840#discussion_r723464012) to add a new nob for just patching only a specific field via an endpoint of a subresource, without modifying the field itself.

Reviewer 10: ```suggestion
To solve this issue, there is [an idea](https://github.com/kubernetes/enhancements/pull/2840#discussion_r723464012) to add a new knob for just patching only a specific field via an endpoint of a subresource, without modifying the field itself.
```
?

In this way, a feature to set a separate RBAC can be implemented later without making liens field itself a subresource.

> Therefore, liens field isn't defined as a subresource and providing a separate RBAC for liens is done in a separate KEP/PR along with finalizers.

Reviewer 3: I think subresource is a beta requirement

Reviewer 11: I would expect the API shape to be completed while in alpha, not beta

Reviewer 3: I meant "to enter beta" - are you saying we should hold this from entering alpha?

Reviewer 11: > I meant "to enter beta" - are you saying we should hold this from entering alpha?

not necessarily, I just wouldn't want the first time we see the shape of the subresource API / interactions to be in the release that promoted to beta.

Reviewer 11: to clarify: I think the subresource aspect is one of the more important and tricky aspects to get right, so I'd really encourage figuring that out as part of the initial alpha implementation, rather than shipping an alpha to get feedback on an API we know doesn't resemble the final shape

>     - [x] Not always. Create/Update permission for resource or update permission for resource/liens subresource are required (subresourcing is a beta blocker).

Reviewer 6: Flagging to myself.  This is mentioned throughout, but I think it's important for alpha.  If we don't do it in alpha, I lack confidence that we will do it before beta.

Reviewer 3: I wont' ACK a beta without subresource, personally.  I think we don't do enough subresourcing.

	- [ ] ~~No. Only create/update permission forthe resource is required.~~
	- [ ] Not always. Create/Update permission for resource or update permission for resource/liens subresource are required (Current KEP's description (Also see [Race of removing/adding liens](#race-of-removingadding-liens))).
  - To remove a lien?
    - [x] Delete permission for the resource/liens subresource is required.
	- [ ] ~~Delete permission for the resource is required.~~
	- [ ] Delete permission for the resource or delete permission for the resource/liens subresource are required (Current KEP's description).
  - To delete anyway and ignore liens?
    - [x] `--not-block-on-lien` kubectl option will pass `NotBlockOnLien` API option in `DeleteOptions` to force delete.
  - Who can ignore liens and how it can be specified?
    - [x] Users who have delete permission for the resource/liens subresource without any additional configurations (If Liens can only be changed via a call to resource/liens subresource, normal users cannot use it to create undeletable resources).

>     - [ ] Users who are specified by kube-apiserver's command line options, `--allowed-not-block-on-lien-users` and `--allowed-not-block-on-lien-groups` (Current KEP's description).

Reviewer 10: If we add these options, are we baking in the decisions that exemptions are always granted to users (and not groups, or via extra auth data)?

#### Unresolved issues

- Decision: When you delete a namespace and an object in that namespace has a lien:
  - Does the namespace and object get deleted anyway?
    - [x] Yes.
    - [ ] No. See below cases:
      - Does the namespace deletion get blocked (i.e., the delete request fails) if any object in the namespace has a lien?
        - No.
      - Does the other objects in the namespace get deleted, but that not object, preventing the namespace deletion from completing?
        - [x] No. Namespace lifecycle cleanup will not honor liens. if you delete the namespace, the content will be removed.
	    - [ ] Yes. See below case:
          - Does the namespace deletion not proceed until all objects in the namespace are free of liens?
            - [x] No. Namespace lifecycle cleanup will not honor liens. If you delete the namespace, the content will be removed.
	        - [ ] ~~Yes. Namespace lifecycle cleanup will honor liens. If you delete the namespace, it won't be deleted until all the content will be deleted.~~

> 	        - [ ] Partially yes. Namespace lifecycle cleanup will honor liens. If you delete the namespace, it won't start deleting the content with liens until all the content without liens will be deleted (Current KEP's description (Also see [Namespace Deletion](#namespace-deletion))).

Author: I agree that cluster admins should have a power to eventually delete the namespace. However, I would like to make this feature to be able to keep the order of deletion as much as possible. In most use cases, order of deletion is required within a namspace. So, if it isn't kept when deleted via namespace, this feature may become useless. WDYT?

(Please also see [here](https://github.com/kubernetes/enhancements/pull/2840#discussion_r714364698) for use case and [here](https://github.com/kubernetes/enhancements/pull/2840#discussion_r715783357) for an idea to solve it).

Reviewer 3: I think this is important, but it's also a niche case and we could go into Alpha an refine it before beta.

Something like: when deleting a namespace, any resource which fails to delete because of a lien is moved to the back of the list of things-to-delete. If we find that the list only includes things that fail, we leave the NS deletion pending.

- Name:
    - [x] "liens" is short and precise, but an unusual word
    - [ ] Other candidates: InUse, deleteInhibitors, hold, lease, claim, deletionBlockers, protections, guards
- Subresource:
  - Do you need a special permission to add a lien?

>     - [x] Yes. Liens can only be changed via a call to resource/liens subresource. Create/update permission for the resource/liens subresource is required.

Author: I may be asking a stupid question, though.
Can we just add liens field as a subresource in metadata, technically?

(I was thinking that liens also need to be subresourced in the same way to "subresource finalizer" KEP. And if so, it would be difficult to rely on the delete permission to liens subresource).
